diff --git a/kernel/Kconfig b/kernel/Kconfig
index 9e3be0b12ab..8206927169c 100644
--- a/kernel/Kconfig
+++ b/kernel/Kconfig
@@ -24,6 +24,12 @@ config MULTITHREADING
 	  set to 'n'; disable only when you REALLY know what you are
 	  doing.
 
+config SYS_THREAD
+	bool "Use common thread compatibility options"
+	default y
+	help
+	  Use common thread compatibility options
+
 config NUM_COOP_PRIORITIES
 	int "Number of coop priorities" if MULTITHREADING
 	default 1 if !MULTITHREADING
