diff --git a/doc/services/shell/index.rst b/doc/services/shell/index.rst
index 77b891175f7..f55b40513f3 100644
--- a/doc/services/shell/index.rst
+++ b/doc/services/shell/index.rst
@@ -714,21 +714,21 @@ An example thread safe usage:
 .. code-block:: c
 
   char *cvalue = NULL;
-  struct getopt_state *state;
-  while ((char c = getopt(argc, argv, "abhc:")) != -1) {
-        state = getopt_state_get();
+  char *optarg;
+  int opterr;
+  int optind = 1;
+  int optopt;
+
+  while ((char c = getopt_r(argc, argv, "abhc:", &optarg, &opterr, &optind, &optopt)) != -1) {
         switch (c) {
         case 'c':
-                cvalue = state->optarg;
+                cvalue = optarg;
                 break;
         default:
                 break;
         }
   }
 
-Thread safe getopt functionality is activated by
-:kconfig:option:`CONFIG_SHELL_GETOPT` set to ``y``.
-
 Obscured Input Feature
 **********************
 
diff --git a/include/zephyr/shell/shell.h b/include/zephyr/shell/shell.h
index ba275c5393c..896e65dc2ae 100644
--- a/include/zephyr/shell/shell.h
+++ b/include/zephyr/shell/shell.h
@@ -19,10 +19,6 @@
 #include <zephyr/sys/util.h>
 #include <zephyr/toolchain.h>
 
-#if defined CONFIG_SHELL_GETOPT
-#include <getopt.h>
-#endif
-
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -993,11 +989,6 @@ struct shell_ctx {
 	/*!< Logging level for a backend. */
 	uint32_t log_level;
 
-#if defined CONFIG_SHELL_GETOPT
-	/*!< getopt context for a shell backend. */
-	struct getopt_state getopt;
-#endif
-
 	uint16_t cmd_buff_len; /*!< Command length.*/
 	uint16_t cmd_buff_pos; /*!< Command buffer cursor position.*/
 
diff --git a/samples/shields/npm6001_ek/prj.conf b/samples/shields/npm6001_ek/prj.conf
index b0292200688..751265dfcf4 100644
--- a/samples/shields/npm6001_ek/prj.conf
+++ b/samples/shields/npm6001_ek/prj.conf
@@ -2,9 +2,8 @@
 # SPDX-License-Identifier: Apache-2.0
 
 CONFIG_SHELL=y
-CONFIG_SHELL_GETOPT=y
 CONFIG_POSIX_C_LIB_EXT=y
-CONFIG_GETOPT_LONG=y
+CONFIG_ZEPHYR_GETOPT_LONG=y
 CONFIG_GPIO=y
 CONFIG_REGULATOR=y
 CONFIG_REGULATOR_SHELL=y
diff --git a/samples/subsys/shell/shell_module/prj_getopt.conf b/samples/subsys/shell/shell_module/prj_getopt.conf
index 4d3abc9e3b4..b1afc52993f 100644
--- a/samples/subsys/shell/shell_module/prj_getopt.conf
+++ b/samples/subsys/shell/shell_module/prj_getopt.conf
@@ -1,7 +1,7 @@
 CONFIG_PRINTK=y
 CONFIG_SHELL=y
-CONFIG_SHELL_GETOPT=y
 CONFIG_LOG=y
+CONFIG_POSIX_C_LIB_EXT=y
 CONFIG_INIT_STACKS=y
 CONFIG_THREAD_STACK_INFO=y
 CONFIG_KERNEL_SHELL=y
diff --git a/samples/subsys/shell/shell_module/src/main.c b/samples/subsys/shell/shell_module/src/main.c
index 8f58334f2fe..4cfbc193c60 100644
--- a/samples/subsys/shell/shell_module/src/main.c
+++ b/samples/subsys/shell/shell_module/src/main.c
@@ -4,19 +4,18 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#define _POSIX_C_SOURCE 200809L
+
+#include <ctype.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <zephyr/getopt.h>
 #include <zephyr/kernel.h>
 #include <zephyr/shell/shell.h>
 #include <zephyr/version.h>
 #include <zephyr/logging/log.h>
-#include <stdlib.h>
 #include <zephyr/drivers/uart.h>
-#include <ctype.h>
-
-#ifdef CONFIG_ARCH_POSIX
-#include <unistd.h>
-#else
-#include <unistd.h>
-#endif
 
 LOG_MODULE_REGISTER(app);
 
@@ -104,18 +101,19 @@ static int cmd_demo_board(const struct shell *sh, size_t argc, char **argv)
 	return 0;
 }
 
-#if defined CONFIG_SHELL_GETOPT
-/* Thread save usage */
+/* Thread safe usage */
 static int cmd_demo_getopt_ts(const struct shell *sh, size_t argc,
 			      char **argv)
 {
-	struct getopt_state *state;
 	int aflag = 0;
 	int bflag = 0;
 	int c;
+	char *optarg;
+	int opterr;
+	int optind = 1;
+	int optopt;
 
-	while ((c = getopt(argc, argv, "abhc:")) != -1) {
-		state = getopt_state_get();
+	while ((c = getopt_r(argc, argv, "abhc:", &optarg, &opterr, &optind, &optopt)) != -1) {
 		switch (c) {
 		case 'a':
 			aflag = 1;
@@ -133,18 +131,18 @@ static int cmd_demo_getopt_ts(const struct shell *sh, size_t argc,
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		case '?':
-			if (state->optopt == 'c') {
+			if (optopt == 'c') {
 				shell_print(sh,
 					"Option -%c requires an argument.",
-					state->optopt);
-			} else if (isprint(state->optopt) != 0) {
+					optopt);
+			} else if (isprint(optopt) != 0) {
 				shell_print(sh,
 					"Unknown option `-%c'.",
-					state->optopt);
+					optopt);
 			} else {
 				shell_print(sh,
 					"Unknown option character `\\x%x'.",
-					state->optopt);
+					optopt);
 			}
 			return 1;
 		default:
@@ -163,6 +161,7 @@ static int cmd_demo_getopt(const struct shell *sh, size_t argc,
 	int bflag = 0;
 	int c;
 
+	optind = 1;
 	while ((c = getopt(argc, argv, "abhc:")) != -1) {
 		switch (c) {
 		case 'a':
@@ -202,7 +201,6 @@ static int cmd_demo_getopt(const struct shell *sh, size_t argc,
 	shell_print(sh, "aflag = %d, bflag = %d", aflag, bflag);
 	return 0;
 }
-#endif
 
 static int cmd_demo_params(const struct shell *sh, size_t argc, char **argv)
 {
@@ -325,14 +323,12 @@ SHELL_STATIC_SUBCMD_SET_CREATE(sub_demo,
 	SHELL_CMD(params, NULL, "Print params command.", cmd_demo_params),
 	SHELL_CMD(ping, NULL, "Ping command.", cmd_demo_ping),
 	SHELL_CMD(board, NULL, "Show board name command.", cmd_demo_board),
-#if defined CONFIG_SHELL_GETOPT
 	SHELL_CMD(getopt_thread_safe, NULL,
 		  "Cammand using getopt in thread safe way"
 		  " looking for: \"abhc:\".",
 		  cmd_demo_getopt_ts),
 	SHELL_CMD(getopt, NULL, "Cammand using getopt in non thread safe way"
 		  " looking for: \"abhc:\".\n", cmd_demo_getopt),
-#endif
 	SHELL_SUBCMD_SET_END /* Array terminated. */
 );
 SHELL_CMD_REGISTER(demo, &sub_demo, "Demo commands", NULL);
diff --git a/subsys/net/l2/Kconfig b/subsys/net/l2/Kconfig
index abc17725d07..77180688489 100644
--- a/subsys/net/l2/Kconfig
+++ b/subsys/net/l2/Kconfig
@@ -71,8 +71,6 @@ config NET_L2_WIFI_SHELL
 	bool "Wi-Fi shell module"
 	select NET_L2_WIFI_MGMT
 	select NET_L2_WIFI_UTILS
-	select SHELL_GETOPT
-	select GETOPT_LONG
 	help
 	  This can be used for controlling Wi-Fi through the console via
 	  exposing a shell module named "wifi".
diff --git a/subsys/net/l2/wifi/wifi_shell.c b/subsys/net/l2/wifi/wifi_shell.c
--- a/subsys/net/l2/wifi/wifi_shell.c
+++ b/subsys/net/l2/wifi/wifi_shell.c
@@ -14,6 +14,7 @@
 
 #include <zephyr/kernel.h>
 #include <getopt.h>
+#include <zephyr/getopt.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -583,7 +584,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"ssid", required_argument, 0, 's'},
 		{"passphrase", required_argument, 0, 'p'},
@@ -648,12 +652,13 @@
 	params->bandwidth = WIFI_FREQ_BANDWIDTH_20MHZ;
 	params->verify_peer_cert = false;
 
-	while ((opt = getopt_long(argc, argv, "s:p:k:e:w:b:c:m:t:a:B:K:S:T:A:V:I:P:g:Rh:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "s:p:k:e:w:b:c:m:t:a:B:K:S:T:A:V:I:P:g:Rh:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 's':
-			params->ssid = state->optarg;
+			params->ssid = optarg;
 			params->ssid_length = strlen(params->ssid);
 			if (params->ssid_length > WIFI_SSID_MAX_LEN) {
 				PR_WARNING("SSID too long (max %d characters)\n",
@@ -662,17 +667,17 @@
 			}
 			break;
 		case 'k':
-			params->security = atoi(state->optarg);
+			params->security = atoi(optarg);
 			if (params->security) {
 				secure_connection = true;
 			}
 			break;
 		case 'p':
-			params->psk = state->optarg;
+			params->psk = optarg;
 			params->psk_length = strlen(params->psk);
 			break;
 		case 'c':
-			channel = strtol(state->optarg, &endptr, 10);
+			channel = strtol(optarg, &endptr, 10);
 			if (iface_mode == WIFI_MODE_AP && channel == 0) {
 				params->channel = channel;
 				break;
@@ -709,7 +714,7 @@
 		case 'b':
 			if (iface_mode == WIFI_MODE_INFRA ||
 			    iface_mode == WIFI_MODE_AP) {
-				switch (atoi(state->optarg)) {
+				switch (atoi(optarg)) {
 				case 2:
 					params->band = WIFI_FREQ_BAND_2_4_GHZ;
 					break;
@@ -727,7 +732,7 @@
 					}
 					__fallthrough;
 				default:
-					PR_ERROR("Invalid band: %d\n", atoi(state->optarg));
+					PR_ERROR("Invalid band: %d\n", atoi(optarg));
 					return -EINVAL;
 				}
 			}
@@ -739,26 +744,26 @@
 					 wifi_security_txt(params->security));
 				return -EINVAL;
 			}
-			params->mfp = atoi(state->optarg);
+			params->mfp = atoi(optarg);
 			break;
 		case 'm':
 			if (net_bytes_from_str(params->bssid, sizeof(params->bssid),
-					       state->optarg) < 0) {
+					       optarg) < 0) {
 				PR_WARNING("Invalid MAC address\n");
 				return -EINVAL;
 			}
 			break;
 		case 't':
 			if (iface_mode == WIFI_MODE_INFRA) {
-				params->timeout = strtol(state->optarg, &endptr, 10);
+				params->timeout = strtol(optarg, &endptr, 10);
 				if (*endptr != '\0') {
-					PR_ERROR("Invalid timeout: %s\n", state->optarg);
+					PR_ERROR("Invalid timeout: %s\n", optarg);
 					return -EINVAL;
 				}
 			}
 			break;
 		case 'a':
-			params->anon_id = state->optarg;
+			params->anon_id = optarg;
 			params->aid_length = strlen(params->anon_id);
 			if (params->aid_length > WIFI_ENT_IDENTITY_MAX_LEN) {
 				PR_WARNING("anon_id too long (max %d characters)\n",
@@ -768,7 +773,7 @@
 			break;
 		case 'B':
 			if (iface_mode == WIFI_MODE_AP) {
-				switch (atoi(state->optarg)) {
+				switch (atoi(optarg)) {
 				case 1:
 					params->bandwidth = WIFI_FREQ_BANDWIDTH_20MHZ;
 					break;
@@ -779,7 +784,7 @@
 					params->bandwidth = WIFI_FREQ_BANDWIDTH_80MHZ;
 					break;
 				default:
-					PR_ERROR("Invalid bandwidth: %d\n", atoi(state->optarg));
+					PR_ERROR("Invalid bandwidth: %d\n", atoi(optarg));
 					return -EINVAL;
 				}
 			}
@@ -791,7 +796,7 @@
 			}
 
 			if (key_passwd_cnt == 0) {
-				params->key_passwd = state->optarg;
+				params->key_passwd = optarg;
 				params->key_passwd_length = strlen(params->key_passwd);
 				if (params->key_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 					PR_WARNING("key_passwd too long (max %d characters)\n",
@@ -799,7 +804,7 @@
 					return -EINVAL;
 				}
 			} else if (key_passwd_cnt == 1) {
-				params->key2_passwd = state->optarg;
+				params->key2_passwd = optarg;
 				params->key2_passwd_length = strlen(params->key2_passwd);
 				if (params->key2_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 					PR_WARNING("key2_passwd too long (max %d characters)\n",
@@ -810,18 +815,18 @@
 			key_passwd_cnt++;
 			break;
 		case 'S':
-			params->wpa3_ent_mode = atoi(state->optarg);
+			params->wpa3_ent_mode = atoi(optarg);
 			break;
 		case 'T':
-			params->TLS_cipher = atoi(state->optarg);
+			params->TLS_cipher = atoi(optarg);
 			break;
 		case 'A':
 			if (iface_mode == WIFI_MODE_INFRA) {
-				params->verify_peer_cert = !!atoi(state->optarg);
+				params->verify_peer_cert = !!atoi(optarg);
 			}
 			break;
 		case 'V':
-			params->eap_ver = atoi(state->optarg);
+			params->eap_ver = atoi(optarg);
 			if (params->eap_ver != 0U && params->eap_ver != 1U) {
 				PR_WARNING("eap_ver error %d\n", params->eap_ver);
 				return -EINVAL;
@@ -834,10 +839,10 @@
 				return -EINVAL;
 			}
 
-			params->eap_identity = state->optarg;
+			params->eap_identity = optarg;
 			params->eap_id_length = strlen(params->eap_identity);
 
-			params->identities[params->nusers] = state->optarg;
+			params->identities[params->nusers] = optarg;
 			params->nusers++;
 			if (params->eap_id_length > WIFI_ENT_IDENTITY_MAX_LEN) {
 				PR_WARNING("eap identity too long (max %d characters)\n",
@@ -852,10 +857,10 @@
 				return -EINVAL;
 			}
 
-			params->eap_password = state->optarg;
+			params->eap_password = optarg;
 			params->eap_passwd_length = strlen(params->eap_password);
 
-			params->passwords[params->passwds] = state->optarg;
+			params->passwords[params->passwds] = optarg;
 			params->passwds++;
 			if (params->eap_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 				PR_WARNING("eap password length too long (max %d characters)\n",
@@ -867,7 +872,7 @@
 			params->ft_used = true;
 			break;
 		case 'g':
-			params->ignore_broadcast_ssid = shell_strtol(state->optarg, 10, &ret);
+			params->ignore_broadcast_ssid = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
@@ -875,7 +880,7 @@
 		case 'h':
 			return -ENOEXEC;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -1017,7 +1022,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"type", required_argument, 0, 't'},
 		{"bands", required_argument, 0, 'b'},
@@ -1034,24 +1042,25 @@
 
 	*do_scan = true;
 
-	while ((opt = getopt_long(argc, argv, "t:b:a:p:s:m:c:i:h",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "t:b:a:p:s:m:c:i:h",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 't':
-			if (!strncasecmp(state->optarg, "passive", 7)) {
+			if (!strncasecmp(optarg, "passive", 7)) {
 				params->scan_type = WIFI_SCAN_TYPE_PASSIVE;
-			} else if (!strncasecmp(state->optarg, "active", 6)) {
+			} else if (!strncasecmp(optarg, "active", 6)) {
 				params->scan_type = WIFI_SCAN_TYPE_ACTIVE;
 			} else {
-				PR_ERROR("Invalid scan type %s\n", state->optarg);
+				PR_ERROR("Invalid scan type %s\n", optarg);
 				return -ENOEXEC;
 			}
 
 			opt_num++;
 			break;
 		case 'b':
-			if (wifi_utils_parse_scan_bands(state->optarg, &params->bands)) {
+			if (wifi_utils_parse_scan_bands(optarg, &params->bands)) {
 				PR_ERROR("Invalid band value(s)\n");
 				return -ENOEXEC;
 			}
@@ -1059,7 +1068,7 @@
 			opt_num++;
 			break;
 		case 'a':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 
 			if (val < 0) {
 				PR_ERROR("Invalid dwell_time_active val\n");
@@ -1070,7 +1079,7 @@
 			opt_num++;
 			break;
 		case 'p':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 
 			if (val < 0) {
 				PR_ERROR("Invalid dwell_time_passive val\n");
@@ -1081,7 +1090,7 @@
 			opt_num++;
 			break;
 		case 's':
-			if (wifi_utils_parse_scan_ssids(state->optarg,
+			if (wifi_utils_parse_scan_ssids(optarg,
 							params->ssids,
 							ARRAY_SIZE(params->ssids))) {
 				PR_ERROR("Invalid SSID(s)\n");
@@ -1091,7 +1100,7 @@
 			opt_num++;
 			break;
 		case 'm':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 
 			if ((val < 0) || (val > WIFI_MGMT_SCAN_MAX_BSS_CNT)) {
 				PR_ERROR("Invalid max_bss val\n");
@@ -1102,7 +1111,7 @@
 			opt_num++;
 			break;
 		case 'c':
-			if (wifi_utils_parse_scan_chan(state->optarg,
+			if (wifi_utils_parse_scan_chan(optarg,
 						       params->band_chan,
 						       ARRAY_SIZE(params->band_chan))) {
 				PR_ERROR("Invalid band or channel value(s)\n");
@@ -1729,7 +1738,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	long value;
 	double twt_mantissa_scale = 0.0;
 	double twt_interval_scale = 0.0;
@@ -1756,12 +1768,13 @@
 
 	params->operation = WIFI_TWT_SETUP;
 
-	while ((opt = getopt_long(argc, argv, "n:c:t:f:r:T:I:a:t:w:p:D:d:e:m:i:h",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "n:c:t:f:r:T:I:a:t:w:p:D:d:e:m:i:h",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'n':
-			if (!parse_number(sh, &value, state->optarg, NULL,
+			if (!parse_number(sh, &value, optarg, NULL,
 					  WIFI_TWT_INDIVIDUAL,
 					  WIFI_TWT_WAKE_TBTT)) {
 				return -EINVAL;
@@ -1770,7 +1783,7 @@
 			break;
 
 		case 'c':
-			if (!parse_number(sh, &value, state->optarg, NULL,
+			if (!parse_number(sh, &value, optarg, NULL,
 					  WIFI_TWT_SETUP_CMD_REQUEST,
 					  WIFI_TWT_SETUP_CMD_DEMAND)) {
 				return -EINVAL;
@@ -1779,14 +1792,14 @@
 			break;
 
 		case 't':
-			if (!parse_number(sh, &value, state->optarg, NULL, 1, 255)) {
+			if (!parse_number(sh, &value, optarg, NULL, 1, 255)) {
 				return -EINVAL;
 			}
 			params->dialog_token = (uint8_t)value;
 			break;
 
 		case 'f':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			if (!parse_number(sh, &value, optarg, NULL, 0,
 					  (WIFI_MAX_TWT_FLOWS - 1))) {
 				return -EINVAL;
 			}
@@ -1794,35 +1807,35 @@
 			break;
 
 		case 'r':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.responder = (bool)value;
 			break;
 
 		case 'T':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.trigger = (bool)value;
 			break;
 
 		case 'I':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.implicit = (bool)value;
 			break;
 
 		case 'a':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.announce = (bool)value;
 			break;
 
 		case 'w':
-			if (!parse_number(sh, &value, state->optarg, NULL, 1,
+			if (!parse_number(sh, &value, optarg, NULL, 1,
 					  WIFI_MAX_TWT_WAKE_INTERVAL_US)) {
 				return -EINVAL;
 			}
@@ -1830,7 +1843,7 @@
 			break;
 
 		case 'p':
-			if (!parse_number(sh, &value, state->optarg, NULL, 1,
+			if (!parse_number(sh, &value, optarg, NULL, 1,
 					  WIFI_MAX_TWT_INTERVAL_US)) {
 				return -EINVAL;
 			}
@@ -1838,7 +1851,7 @@
 			break;
 
 		case 'D':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			if (!parse_number(sh, &value, optarg, NULL, 0,
 					  WIFI_MAX_TWT_WAKE_AHEAD_DURATION_US)) {
 				return -EINVAL;
 			}
@@ -1846,14 +1859,14 @@
 			break;
 
 		case 'd':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.twt_info_disable = (bool)value;
 			break;
 
 		case 'e':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			if (!parse_number(sh, &value, optarg, NULL, 0,
 					  WIFI_MAX_TWT_EXPONENT)) {
 				return -EINVAL;
 			}
@@ -1861,7 +1874,7 @@
 			break;
 
 		case 'm':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 0xFFFF)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 0xFFFF)) {
 				return -EINVAL;
 			}
 			params->setup.twt_mantissa = (uint16_t)value;
@@ -2138,7 +2151,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"max_inactivity", required_argument, 0, 't'},
 		{"max_num_sta", required_argument, 0, 's'},
@@ -2151,12 +2167,13 @@
 		{0, 0, 0, 0}};
 	long val;
 
-	while ((opt = getopt_long(argc, argv, "t:s:n:c:i:h",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "t:s:n:c:i:h",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 't':
-			if (!parse_number(sh, &val, state->optarg, "max_inactivity",
+			if (!parse_number(sh, &val, optarg, "max_inactivity",
 					  0, WIFI_AP_STA_MAX_INACTIVITY)) {
 				return -EINVAL;
 			}
@@ -2164,7 +2181,7 @@
 			params->type |= WIFI_AP_CONFIG_PARAM_MAX_INACTIVITY;
 			break;
 		case 's':
-			if (!parse_number(sh, &val, state->optarg, "max_num_sta",
+			if (!parse_number(sh, &val, optarg, "max_num_sta",
 					  0, CONFIG_WIFI_MGMT_AP_MAX_NUM_STA)) {
 				return -EINVAL;
 			}
@@ -2173,11 +2190,11 @@
 			break;
 #if defined(CONFIG_WIFI_NM_HOSTAPD_AP)
 		case 'n':
-			strncpy(params->ht_capab, state->optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
+			strncpy(params->ht_capab, optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
 			params->type |= WIFI_AP_CONFIG_PARAM_HT_CAPAB;
 			break;
 		case 'c':
-			strncpy(params->vht_capab, state->optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
+			strncpy(params->vht_capab, optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
 			params->type |= WIFI_AP_CONFIG_PARAM_VHT_CAPAB;
 			break;
 #endif
@@ -2188,7 +2205,7 @@
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		}
@@ -2258,19 +2275,24 @@
 static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
 				   char *argv[])
 {
 	struct net_if *iface = get_iface(IFACE_TYPE_STA, argc, argv);
 	struct wifi_reg_domain regd = {0};
 	int ret, chan_idx = 0;
 	int opt;
 	bool force = false;
 	bool verbose = false;
 	int opt_index = 0;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"force", no_argument, 0, 'f'},
 		{"verbose", no_argument, 0, 'v'},
 		{"iface", required_argument, 0, 'i'},
 		{NULL, 0, NULL, 0}
 	};
 
-	while ((opt = getopt_long(argc, argv, "fvi:", long_options, &opt_index)) != -1) {
+	while ((opt = getopt_long_r(argc, argv, "fvi:", long_options, &opt_index, &optarg,
+				    &opterr, &optind, &optopt)) != -1) {
 		switch (opt) {
@@ -2635,7 +2652,10 @@
 	int opt;
 	int opt_index = 0;
 	int opt_num = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"iface", required_argument, 0, 'i'},
 		{"sta", no_argument, 0, 's'},
@@ -2646,9 +2666,10 @@
 		{0, 0, 0, 0}};
 
 	mode->oper = WIFI_MGMT_GET;
-	while ((opt = getopt_long(argc, argv, "i:smtpakh",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "i:smtpakh",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 's':
 			mode->mode |= WIFI_STA_MODE;
@@ -2667,7 +2688,7 @@
 			opt_num++;
 			break;
 		case 'i':
-			mode->if_index = (uint8_t)atoi(state->optarg);
+			mode->if_index = (uint8_t)atoi(optarg);
 			/* Don't count iface as it's common for both get and set */
 			break;
 		case 'h':
@@ -2738,7 +2759,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"iface", optional_argument, 0, 'i'},
 		{"channel", required_argument, 0, 'c'},
@@ -2746,15 +2770,16 @@
 		{"help", no_argument, 0, 'h'},
 		{0, 0, 0, 0}};
 
-	while ((opt = getopt_long(argc, argv, "i:c:gh",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "i:c:gh",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'c':
-			channel->channel = (uint16_t)atoi(state->optarg);
+			channel->channel = (uint16_t)atoi(optarg);
 			break;
 		case 'i':
-			channel->if_index = (uint8_t)atoi(state->optarg);
+			channel->if_index = (uint8_t)atoi(optarg);
 			break;
 		case 'g':
 			channel->oper = WIFI_MGMT_GET;
@@ -2836,7 +2861,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"iface", required_argument, 0, 'i'},
 		{"capture-len", optional_argument, 0, 'b'},
@@ -2848,9 +2876,10 @@
 		{"help", no_argument, 0, 'h'},
 		{0, 0, 0, 0}};
 
-	while ((opt = getopt_long(argc, argv, "i:b:amcdgh",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "i:b:amcdgh",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'a':
 			filter->filter |= WIFI_PACKET_FILTER_ALL;
@@ -2865,10 +2894,10 @@
 			filter->filter |= WIFI_PACKET_FILTER_DATA;
 			break;
 		case 'i':
-			filter->if_index = (uint8_t)atoi(state->optarg);
+			filter->if_index = (uint8_t)atoi(optarg);
 			break;
 		case 'b':
-			filter->buffer_size = (uint16_t)atoi(state->optarg);
+			filter->buffer_size = (uint16_t)atoi(optarg);
 			break;
 		case 'h':
 			shell_help(sh);
@@ -2960,7 +2989,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"peer", required_argument, 0, 'p'},
 		{"role", required_argument, 0, 'r'},
@@ -2971,30 +3003,31 @@
 		{0, 0, 0, 0}};
 	int ret = 0;
 
-	while ((opt = getopt_long(argc, argv, "p:r:c:m:s:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "p:r:c:m:s:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'p':
-			params->auth_init.peer = shell_strtol(state->optarg, 10, &ret);
+			params->auth_init.peer = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'r':
-			params->auth_init.role = shell_strtol(state->optarg, 10, &ret);
+			params->auth_init.role = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'c':
-			params->auth_init.configurator = shell_strtol(state->optarg, 10, &ret);
+			params->auth_init.configurator = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'm':
-			params->auth_init.conf = shell_strtol(state->optarg, 10, &ret);
+			params->auth_init.conf = shell_strtol(optarg, 10, &ret);
 			break;
 		case 's':
-			strncpy(params->auth_init.ssid, state->optarg, WIFI_SSID_MAX_LEN);
+			strncpy(params->auth_init.ssid, optarg, WIFI_SSID_MAX_LEN);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -3012,7 +3045,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"own", required_argument, 0, 'o'},
 		{"freq", required_argument, 0, 'f'},
@@ -3020,21 +3056,22 @@
 		{0, 0, 0, 0}};
 	int ret = 0;
 
-	while ((opt = getopt_long(argc, argv, "o:f:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "o:f:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'o':
-			params->chirp.id = shell_strtol(state->optarg, 10, &ret);
+			params->chirp.id = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'f':
-			params->chirp.freq = shell_strtol(state->optarg, 10, &ret);
+			params->chirp.freq = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -3052,7 +3089,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"role", required_argument, 0, 'r'},
 		{"freq", required_argument, 0, 'f'},
@@ -3060,21 +3100,22 @@
 		{0, 0, 0, 0}};
 	int ret = 0;
 
-	while ((opt = getopt_long(argc, argv, "r:f:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "r:f:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'r':
-			params->listen.role = shell_strtol(state->optarg, 10, &ret);
+			params->listen.role = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'f':
-			params->listen.freq = shell_strtol(state->optarg, 10, &ret);
+			params->listen.freq = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -3092,7 +3133,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"type", required_argument, 0, 't'},
 		{"opclass", required_argument, 0, 'o'},
@@ -3102,28 +3146,29 @@
 		{0, 0, 0, 0}};
 	int ret = 0;
 
-	while ((opt = getopt_long(argc, argv, "t:o:h:a:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "t:o:h:a:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 't':
-			params->bootstrap_gen.type = shell_strtol(state->optarg, 10, &ret);
+			params->bootstrap_gen.type = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'o':
-			params->bootstrap_gen.op_class = shell_strtol(state->optarg, 10, &ret);
+			params->bootstrap_gen.op_class = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'h':
-			params->bootstrap_gen.chan = shell_strtol(state->optarg, 10, &ret);
+			params->bootstrap_gen.chan = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'a':
 			ret = net_bytes_from_str(params->bootstrap_gen.mac,
-						 WIFI_MAC_ADDR_LEN, state->optarg);
+						 WIFI_MAC_ADDR_LEN, optarg);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -3158,7 +3203,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"configurator", required_argument, 0, 'c'},
 		{"mode", required_argument, 0, 'm'},
@@ -3167,25 +3215,26 @@
 		{0, 0, 0, 0}};
 	int ret = 0;
 
-	while ((opt = getopt_long(argc, argv, "p:r:c:m:s:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "p:r:c:m:s:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'c':
 			params->configurator_set.configurator =
-				shell_strtol(state->optarg, 10, &ret);
+				shell_strtol(optarg, 10, &ret);
 			break;
 		case 'm':
-			params->configurator_set.conf = shell_strtol(state->optarg, 10, &ret);
+			params->configurator_set.conf = shell_strtol(optarg, 10, &ret);
 			break;
 		case 's':
-			strncpy(params->configurator_set.ssid, state->optarg, WIFI_SSID_MAX_LEN);
+			strncpy(params->configurator_set.ssid, optarg, WIFI_SSID_MAX_LEN);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -3450,7 +3499,10 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"peer", required_argument, 0, 'p'},
 		{"iface", required_argument, 0, 'i'},
@@ -3461,18 +3513,19 @@
 
 	params.action = WIFI_DPP_AUTH_INIT;
 
-	while ((opt = getopt_long(argc, argv, "p:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "p:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'p':
-			params.auth_init.peer = shell_strtol(state->optarg, 10, &ret);
+			params.auth_init.peer = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 
@@ -3563,7 +3616,10 @@
 	int err;
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"type", required_argument, 0, 't'},
 		{"short-interval", required_argument, 0, 's'},
@@ -3575,50 +3631,51 @@
 	unsigned long uval;
 	long val;
 
-	while ((opt = getopt_long(argc, argv, "t:s:r:l:b:i:", long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "t:s:r:l:b:i:", long_options, &opt_index,
+				    &optarg, &opterr, &optind, &optopt)) != -1) {
+
 		switch (opt) {
 		case 't':
-			if (strcmp("simple", state->optarg) == 0) {
+			if (strcmp("simple", optarg) == 0) {
 				params->type = WIFI_BGSCAN_SIMPLE;
-			} else if (strcmp("learn", state->optarg) == 0) {
+			} else if (strcmp("learn", optarg) == 0) {
 				params->type = WIFI_BGSCAN_LEARN;
-			} else if (strcmp("none", state->optarg) == 0) {
+			} else if (strcmp("none", optarg) == 0) {
 				params->type = WIFI_BGSCAN_NONE;
 			} else {
-				PR_ERROR("Invalid type %s\n", state->optarg);
+				PR_ERROR("Invalid type %s\n", optarg);
 				shell_help(sh);
 				return SHELL_CMD_HELP_PRINTED;
 			}
 			break;
 		case 's':
-			uval = shell_strtoul(state->optarg, 10, &err);
+			uval = shell_strtoul(optarg, 10, &err);
 			if (err < 0) {
-				PR_ERROR("Invalid short interval %s\n", state->optarg);
+				PR_ERROR("Invalid short interval %s\n", optarg);
 				return err;
 			}
 			params->short_interval = uval;
 			break;
 		case 'l':
-			uval = shell_strtoul(state->optarg, 10, &err);
+			uval = shell_strtoul(optarg, 10, &err);
 			if (err < 0) {
-				PR_ERROR("Invalid long interval %s\n", state->optarg);
+				PR_ERROR("Invalid long interval %s\n", optarg);
 				return err;
 			}
 			params->long_interval = uval;
 			break;
 		case 'b':
-			uval = shell_strtoul(state->optarg, 10, &err);
+			uval = shell_strtoul(optarg, 10, &err);
 			if (err < 0) {
-				PR_ERROR("Invalid BTM queries %s\n", state->optarg);
+				PR_ERROR("Invalid BTM queries %s\n", optarg);
 				return err;
 			}
 			params->btm_queries = uval;
 			break;
 		case 'r':
-			val = shell_strtol(state->optarg, 10, &err);
+			val = shell_strtol(optarg, 10, &err);
 			if (err < 0) {
-				PR_ERROR("Invalid RSSI threshold %s\n", state->optarg);
+				PR_ERROR("Invalid RSSI threshold %s\n", optarg);
 				return err;
 			}
 			params->rssi_threshold = val;
@@ -3627,7 +3684,7 @@
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		}
@@ -3668,19 +3725,23 @@
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg = NULL;
+	int opterr = 0;
+	int optind = 1;
+	int optopt = 0;
 	static const struct option long_options[] = {
 		{"okc", required_argument, 0, 'o'},
 		{"iface", required_argument, 0, 'i'},
 		{0, 0, 0, 0}};
 	long val;
 
-	while ((opt = getopt_long(argc, argv, "o:i:",
-				  long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "o:i:",
+				    long_options, &opt_index, &optarg, &opterr, &optind,
+				    &optopt)) != -1) {
+
 		switch (opt) {
 		case 'o':
-			if (!parse_number(sh, &val, state->optarg, "okc", 0, 1)) {
+			if (!parse_number(sh, &val, optarg, "okc", 0, 1)) {
 				return -EINVAL;
 			}
 			params->okc = val;
@@ -3690,7 +3751,7 @@
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		}
diff --git a/subsys/net/lib/wifi_credentials/Kconfig b/subsys/net/lib/wifi_credentials/Kconfig
index 1a5671ccd73..0d53c2b1136 100644
--- a/subsys/net/lib/wifi_credentials/Kconfig
+++ b/subsys/net/lib/wifi_credentials/Kconfig
@@ -57,8 +57,6 @@ config WIFI_CREDENTIALS_SHELL
 	bool "Shell commands to manage Wi-Fi credentials"
 	default y
 	depends on SHELL
-	select SHELL_GETOPT
-	select GETOPT_LONG
 	depends on !WIFI_CREDENTIALS_BACKEND_NONE
 
 config WIFI_CREDENTIALS_CONNECT_STORED
diff --git a/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c b/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
index 3966a2404bc..a16f3e42aa7 100644
--- a/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
+++ b/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
@@ -117,7 +117,10 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 {
 	int opt;
 	int opt_index = 0;
-	struct getopt_state *state;
+	char *optarg;
+	int opterr;
+	int optind = 1;
+	int optopt;
 	static const struct option long_options[] = {
 		{"ssid", required_argument, 0, 's'},	 {"passphrase", required_argument, 0, 'p'},
 		{"key-mgmt", required_argument, 0, 'k'}, {"ieee-80211w", required_argument, 0, 'w'},
@@ -138,27 +141,26 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 	long channel;
 	long mfp = WIFI_MFP_OPTIONAL;
 
-	while ((opt = getopt_long(argc, argv, "s:p:k:w:b:c:m:t:a:K:h", long_options, &opt_index)) !=
-	       -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "s:p:k:w:b:c:m:t:a:K:h", long_options, &opt_index,
+				    &optarg, &opterr, &optind, &optopt)) != -1) {
 		switch (opt) {
 		case 's':
-			creds.header.ssid_len = strlen(state->optarg);
+			creds.header.ssid_len = strlen(optarg);
 			if (creds.header.ssid_len > WIFI_SSID_MAX_LEN) {
 				shell_warn(sh, "SSID too long (max %d characters)\n",
 					   WIFI_SSID_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.header.ssid, state->optarg, creds.header.ssid_len);
+			memcpy(creds.header.ssid, optarg, creds.header.ssid_len);
 			break;
 		case 'k':
-			creds.header.type = atoi(state->optarg);
+			creds.header.type = atoi(optarg);
 			if (creds.header.type) {
 				secure_connection = true;
 			}
 			break;
 		case 'p':
-			creds.password_len = strlen(state->optarg);
+			creds.password_len = strlen(optarg);
 			if (creds.password_len < WIFI_PSK_MIN_LEN) {
 				shell_warn(sh, "Passphrase should be minimum %d characters\n",
 					   WIFI_PSK_MIN_LEN);
@@ -169,12 +171,12 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 					   WIFI_PSK_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.password, state->optarg, creds.password_len);
+			memcpy(creds.password, optarg, creds.password_len);
 			break;
 		case 'c':
-			channel = strtol(state->optarg, &endptr, 10);
+			channel = strtol(optarg, &endptr, 10);
 			if (*endptr != '\0') {
-				shell_error(sh, "Invalid channel: %s\n", state->optarg);
+				shell_error(sh, "Invalid channel: %s\n", optarg);
 				return -EINVAL;
 			}
 
@@ -205,7 +207,7 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 			creds.header.channel = channel;
 			break;
 		case 'b':
-			switch (atoi(state->optarg)) {
+			switch (atoi(optarg)) {
 			case 2:
 				creds.header.flags |= WIFI_CREDENTIALS_FLAG_2_4GHz;
 				break;
@@ -216,7 +218,7 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 				creds.header.flags |= WIFI_CREDENTIALS_FLAG_6GHz;
 				break;
 			default:
-				shell_error(sh, "Invalid band: %d\n", atoi(state->optarg));
+				shell_error(sh, "Invalid band: %d\n", atoi(optarg));
 				return -EINVAL;
 			}
 			break;
@@ -227,9 +229,9 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 					    wifi_security_txt(creds.header.type));
 				return -ENOTSUP;
 			}
-			mfp = strtol(state->optarg, &endptr, 10);
+			mfp = strtol(optarg, &endptr, 10);
 			if (*endptr != '\0') {
-				shell_error(sh, "Invalid IEEE 802.11w value: %s", state->optarg);
+				shell_error(sh, "Invalid IEEE 802.11w value: %s", optarg);
 				return -EINVAL;
 			}
 			if (mfp == WIFI_MFP_DISABLE) {
@@ -237,45 +239,43 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 			} else if (mfp == WIFI_MFP_REQUIRED) {
 				creds.header.flags |= WIFI_CREDENTIALS_FLAG_MFP_REQUIRED;
 			} else if (mfp > 2) {
-				shell_error(sh, "Invalid IEEE 802.11w value: %s",
-					    state->optarg);
+				shell_error(sh, "Invalid IEEE 802.11w value: %s", optarg);
 				return -EINVAL;
 			}
 			break;
 		case 'm':
 			if (net_bytes_from_str(creds.header.bssid, sizeof(creds.header.bssid),
-					       state->optarg) < 0) {
+					       optarg) < 0) {
 				shell_warn(sh, "Invalid MAC address\n");
 				return -EINVAL;
 			}
 			creds.header.flags |= WIFI_CREDENTIALS_FLAG_BSSID;
 			break;
 		case 'a':
-			creds.header.aid_length = strlen(state->optarg);
+			creds.header.aid_length = strlen(optarg);
 			if (creds.header.aid_length > WIFI_ENT_IDENTITY_MAX_LEN) {
 				shell_warn(sh, "anon_id too long (max %d characters)\n",
 					   WIFI_ENT_IDENTITY_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.header.anon_id, state->optarg, creds.header.aid_length);
+			memcpy(creds.header.anon_id, optarg, creds.header.aid_length);
 			creds.header.flags |= WIFI_CREDENTIALS_FLAG_ANONYMOUS_IDENTITY;
 			break;
 		case 'K':
-			creds.header.key_passwd_length = strlen(state->optarg);
+			creds.header.key_passwd_length = strlen(optarg);
 			if (creds.header.key_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 				shell_warn(sh, "key_passwd too long (max %d characters)\n",
 					   WIFI_ENT_PSWD_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.header.key_passwd, state->optarg,
-			       creds.header.key_passwd_length);
+			memcpy(creds.header.key_passwd, optarg, creds.header.key_passwd_length);
 			creds.header.flags |= WIFI_CREDENTIALS_FLAG_KEY_PASSWORD;
 			break;
 		case 'h':
 			shell_help(sh);
 			return -ENOEXEC;
 		default:
-			shell_error(sh, "Invalid option %c\n", state->optopt);
+			shell_error(sh, "Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 	}
diff --git a/subsys/shell/Kconfig b/subsys/shell/Kconfig
index c35b0193f6b..a22f1dab341 100644
--- a/subsys/shell/Kconfig
+++ b/subsys/shell/Kconfig
@@ -189,13 +189,10 @@ config SHELL_VT100_COLORS
 	  If enabled VT100 colors are used in shell (e.g. print errors in red).
 
 config SHELL_GETOPT
-	bool "Threadsafe getopt support in shell"
-	select POSIX_C_LIB_EXT
+	bool "Threadsafe getopt support in shell [DEPRECATED]"
+	select DEPRECATED
 	help
-	  This config creates a separate getopt_state for the shell instance.
-	  It ensures that using getopt with shell is thread safe.
-	  When more threads are using getopt please call getopt_state_get to
-	  get getopt state of the shell thread.
+	  This option is deprecated. Please use `getopt_r()` instead of `getopt()`.
 
 config SHELL_METAKEYS
 	bool "Metakeys"
@@ -214,7 +211,6 @@ config SHELL_HELP
 config SHELL_HELP_OPT_PARSE
 	bool "Parse -h and --help options"
 	depends on SHELL_HELP
-	depends on !SHELL_GETOPT
 	default y
 	help
 	  Shell parses command to find '-h' or '--help' string. If the shell
diff --git a/subsys/shell/modules/devmem_service.c b/subsys/shell/modules/devmem_service.c
index df37a18f4d0..bf5adb0b1d3 100644
--- a/subsys/shell/modules/devmem_service.c
+++ b/subsys/shell/modules/devmem_service.c
@@ -9,17 +9,16 @@
 #undef _POSIX_C_SOURCE
 #define _POSIX_C_SOURCE 200809L
 
+#include <getopt.h>
 #include <stdlib.h>
 #include <unistd.h>
+
 #include <zephyr/device.h>
+#include <zephyr/getopt.h>
 #include <zephyr/shell/shell.h>
 #include <zephyr/sys/byteorder.h>
 #include <zephyr/sys/util.h>
 
-#ifndef CONFIG_NATIVE_LIBC
-extern void getopt_init(void);
-#endif
-
 static inline bool is_ascii(uint8_t data)
 {
 	return (data >= 0x30 && data <= 0x39) || (data >= 0x61 && data <= 0x66) ||
@@ -118,11 +117,12 @@ static int cmd_dump(const struct shell *sh, size_t argc, char **argv)
 	size_t width = 32;
 	mem_addr_t addr = -1;
 
-	optind = 1;
-#ifndef CONFIG_NATIVE_LIBC
-	getopt_init();
-#endif
-	while ((rv = getopt(argc, argv, "a:s:w:")) != -1) {
+	char *optarg;
+	int opterr;
+	int optind = 1;
+	int optopt;
+
+	while ((rv = getopt_r(argc, argv, "a:s:w:", &optarg, &opterr, &optind, &optopt)) != -1) {
 		switch (rv) {
 		case 'a':
 			addr = (mem_addr_t)shell_strtoul(optarg, 16, &err);
diff --git a/subsys/shell/shell.c b/subsys/shell/shell.c
index 8c3380d23ba..8bcbb512083 100644
--- a/subsys/shell/shell.c
+++ b/subsys/shell/shell.c
@@ -545,10 +545,6 @@ static int exec_cmd(const struct shell *sh, size_t argc, const char **argv,
 	}
 
 	if (!ret_val) {
-#if CONFIG_SHELL_GETOPT
-		getopt_init();
-#endif
-
 		z_flag_cmd_ctx_set(sh, true);
 		/* Unlock thread mutex in case command would like to borrow
 		 * shell context to other thread to avoid mutex deadlock.
diff --git a/subsys/testsuite/ztest/Kconfig b/subsys/testsuite/ztest/Kconfig
index 5f07f448103..d1a15b058a7 100644
--- a/subsys/testsuite/ztest/Kconfig
+++ b/subsys/testsuite/ztest/Kconfig
@@ -27,8 +27,6 @@ config ZTEST_SHELL
 	bool "Ztest with shell support"
 	select SHELL
 	select SHELL_THREAD_PRIORITY_OVERRIDE
-	select GETOPT_LONG
-	select SHELL_GETOPT
 	help
 	  Enable shell to manage test execution and selection.
 
diff --git a/subsys/testsuite/ztest/src/ztest.c b/subsys/testsuite/ztest/src/ztest.c
index 080e1726528..bae835fbaf2 100644
--- a/subsys/testsuite/ztest/src/ztest.c
+++ b/subsys/testsuite/ztest/src/ztest.c
@@ -1253,7 +1253,6 @@ static int cmd_runall(const struct shell *sh, size_t argc, char **argv)
 static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
 {
 
-	struct getopt_state *state;
 	int opt;
 	static struct option long_options[] = {{"suite_iter", required_argument, 0, 's'},
 					       {"case_iter", required_argument, 0, 'c'},
@@ -1265,11 +1264,16 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
 	int suite_iter = 1;
 	int case_iter = 1;
 
-	while ((opt = getopt_long(argc, argv, "s:c:", long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	char *optarg;
+	int opterr;
+	int optind = 1;
+	int optopt;
+
+	while ((opt = getopt_long_r(argc, argv, "s:c:", long_options, &opt_index, &optarg, &opterr,
+				    &optind, &optopt)) != -1) {
 		switch (opt) {
 		case 's':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 			if (val < 1) {
 				shell_error(sh, "Invalid number of suite iterations");
 				return -ENOEXEC;
@@ -1278,7 +1282,7 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
 			opt_num++;
 			break;
 		case 'c':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 			if (val < 1) {
 				shell_error(sh, "Invalid number of case iterations");
 				return -ENOEXEC;
@@ -1301,7 +1305,6 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
 
 static int cmd_run_suite(const struct shell *sh, size_t argc, char **argv)
 {
-	struct getopt_state *state;
 	int opt;
 	static struct option long_options[] = {{"repeat_iter", required_argument, NULL, 'r'},
 		{NULL, 0, NULL, 0}};
@@ -1310,12 +1313,16 @@ static int cmd_run_suite(const struct shell *sh, size_t argc, char **argv)
 	int opt_num = 0;
 	void *param = NULL;
 	int repeat_iter = 1;
+	char *optarg;
+	int opterr;
+	int optind = 1;
+	int optopt;
 
-	while ((opt = getopt_long(argc, argv, "r:p:", long_options, &opt_index)) != -1) {
-		state = getopt_state_get();
+	while ((opt = getopt_long_r(argc, argv, "r:p:", long_options, &opt_index, &optarg, &opterr,
+				    &optind, &optopt)) != -1) {
 		switch (opt) {
 		case 'r':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 			if (val < 1) {
 				shell_fprintf(sh, SHELL_ERROR,
 					"Invalid number of suite interations\n");
@@ -1325,7 +1332,7 @@ static int cmd_run_suite(const struct shell *sh, size_t argc, char **argv)
 			opt_num++;
 			break;
 		case 'p':
-			param = state->optarg;
+			param = optarg;
 			opt_num++;
 			break;
 		default:
diff --git a/tests/drivers/wifi/nrf_wifi/prj.conf b/tests/drivers/wifi/nrf_wifi/prj.conf
index fb8c6c778a8..58b1e9fd079 100644
--- a/tests/drivers/wifi/nrf_wifi/prj.conf
+++ b/tests/drivers/wifi/nrf_wifi/prj.conf
@@ -53,7 +53,6 @@ CONFIG_REBOOT=y
 CONFIG_LOG=y
 CONFIG_PRINTK=y
 CONFIG_SHELL=y
-CONFIG_SHELL_GETOPT=y
 CONFIG_DEVICE_SHELL=y
 CONFIG_POSIX_TIMERS=y
 CONFIG_DATE_SHELL=y
