diff --git a/samples/shields/npm6001_ek/prj.conf b/samples/shields/npm6001_ek/prj.conf
index b0292200688..751265dfcf4 100644
--- a/samples/shields/npm6001_ek/prj.conf
+++ b/samples/shields/npm6001_ek/prj.conf
@@ -2,9 +2,8 @@
 # SPDX-License-Identifier: Apache-2.0

 CONFIG_SHELL=y
-CONFIG_SHELL_GETOPT=y
 CONFIG_POSIX_C_LIB_EXT=y
-CONFIG_GETOPT_LONG=y
+CONFIG_ZEPHYR_GETOPT_LONG=y
 CONFIG_GPIO=y
 CONFIG_REGULATOR=y
 CONFIG_REGULATOR_SHELL=y
diff --git a/samples/subsys/shell/shell_module/prj_getopt.conf b/samples/subsys/shell/shell_module/prj_getopt.conf
index 4d3abc9e3b4..b1afc52993f 100644
--- a/samples/subsys/shell/shell_module/prj_getopt.conf
+++ b/samples/subsys/shell/shell_module/prj_getopt.conf
@@ -1,7 +1,7 @@
 CONFIG_PRINTK=y
 CONFIG_SHELL=y
-CONFIG_SHELL_GETOPT=y
 CONFIG_LOG=y
+CONFIG_POSIX_C_LIB_EXT=y
 CONFIG_INIT_STACKS=y
 CONFIG_THREAD_STACK_INFO=y
 CONFIG_KERNEL_SHELL=y
diff --git a/samples/subsys/shell/shell_module/src/main.c b/samples/subsys/shell/shell_module/src/main.c
index 674cfb3475d..d3ef16ccec9 100644
--- a/samples/subsys/shell/shell_module/src/main.c
+++ b/samples/subsys/shell/shell_module/src/main.c
@@ -129,18 +129,18 @@ static int cmd_demo_getopt_ts(const struct shell *sh, size_t argc,
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		case '?':
-			if (state->optopt == 'c') {
+			if (optopt == 'c') {
 				shell_print(sh,
 					"Option -%c requires an argument.",
-					state->optopt);
-			} else if (isprint(state->optopt) != 0) {
+					optopt);
+			} else if (isprint(optopt) != 0) {
 				shell_print(sh,
 					"Unknown option `-%c'.",
-					state->optopt);
+					optopt);
 			} else {
 				shell_print(sh,
 					"Unknown option character `\\x%x'.",
-					state->optopt);
+					optopt);
 			}
 			return 1;
 		default:
@@ -200,7 +200,6 @@ static int cmd_demo_getopt(const struct shell *sh, size_t argc,
 	shell_print(sh, "aflag = %d, bflag = %d", aflag, bflag);
 	return 0;
 }
-#endif

 static int cmd_demo_params(const struct shell *sh, size_t argc, char **argv)
 {
diff --git a/subsys/net/l2/Kconfig b/subsys/net/l2/Kconfig
index 64695b54fcf..88edddc39c1 100644
--- a/subsys/net/l2/Kconfig
+++ b/subsys/net/l2/Kconfig
@@ -71,8 +71,6 @@ config NET_L2_WIFI_SHELL
 	bool "Wi-Fi shell module"
 	select NET_L2_WIFI_MGMT
 	select NET_L2_WIFI_UTILS
-	select SHELL_GETOPT
-	select GETOPT_LONG
 	help
 	  This can be used for controlling Wi-Fi through the console via
 	  exposing a shell module named "wifi".
diff --git a/subsys/net/l2/wifi/wifi_shell.c b/subsys/net/l2/wifi/wifi_shell.c
index a194979cb6a..069ef4445b7 100644
--- a/subsys/net/l2/wifi/wifi_shell.c
+++ b/subsys/net/l2/wifi/wifi_shell.c
@@ -764,7 +764,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 		case 'b':
 			if (iface_mode == WIFI_MODE_INFRA ||
 			    iface_mode == WIFI_MODE_AP) {
-				switch (atoi(state->optarg)) {
+				switch (atoi(optarg)) {
 				case 2:
 					params->band = WIFI_FREQ_BAND_2_4_GHZ;
 					break;
@@ -782,7 +782,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 					}
 					__fallthrough;
 				default:
-					PR_ERROR("Invalid band: %d\n", atoi(state->optarg));
+					PR_ERROR("Invalid band: %d\n", atoi(optarg));
 					return -EINVAL;
 				}
 			}
@@ -794,26 +794,26 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 					 wifi_security_txt(params->security));
 				return -EINVAL;
 			}
-			params->mfp = atoi(state->optarg);
+			params->mfp = atoi(optarg);
 			break;
 		case 'm':
 			if (net_bytes_from_str(params->bssid, sizeof(params->bssid),
-					       state->optarg) < 0) {
+					       optarg) < 0) {
 				PR_WARNING("Invalid MAC address\n");
 				return -EINVAL;
 			}
 			break;
 		case 't':
 			if (iface_mode == WIFI_MODE_INFRA) {
-				params->timeout = strtol(state->optarg, &endptr, 10);
+				params->timeout = strtol(optarg, &endptr, 10);
 				if (*endptr != '\0') {
-					PR_ERROR("Invalid timeout: %s\n", state->optarg);
+					PR_ERROR("Invalid timeout: %s\n", optarg);
 					return -EINVAL;
 				}
 			}
 			break;
 		case 'a':
-			params->anon_id = state->optarg;
+			params->anon_id = optarg;
 			params->aid_length = strlen(params->anon_id);
 			if (params->aid_length > WIFI_ENT_IDENTITY_MAX_LEN) {
 				PR_WARNING("anon_id too long (max %d characters)\n",
@@ -823,7 +823,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 			break;
 		case 'B':
 			if (iface_mode == WIFI_MODE_AP) {
-				switch (atoi(state->optarg)) {
+				switch (atoi(optarg)) {
 				case 1:
 					params->bandwidth = WIFI_FREQ_BANDWIDTH_20MHZ;
 					break;
@@ -834,7 +834,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 					params->bandwidth = WIFI_FREQ_BANDWIDTH_80MHZ;
 					break;
 				default:
-					PR_ERROR("Invalid bandwidth: %d\n", atoi(state->optarg));
+					PR_ERROR("Invalid bandwidth: %d\n", atoi(optarg));
 					return -EINVAL;
 				}
 			}
@@ -846,7 +846,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 			}

 			if (key_passwd_cnt == 0) {
-				params->key_passwd = state->optarg;
+				params->key_passwd = optarg;
 				params->key_passwd_length = strlen(params->key_passwd);
 				if (params->key_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 					PR_WARNING("key_passwd too long (max %d characters)\n",
@@ -854,7 +854,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 					return -EINVAL;
 				}
 			} else if (key_passwd_cnt == 1) {
-				params->key2_passwd = state->optarg;
+				params->key2_passwd = optarg;
 				params->key2_passwd_length = strlen(params->key2_passwd);
 				if (params->key2_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 					PR_WARNING("key2_passwd too long (max %d characters)\n",
@@ -892,10 +892,10 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 				return -EINVAL;
 			}

-			params->eap_identity = state->optarg;
+			params->eap_identity = optarg;
 			params->eap_id_length = strlen(params->eap_identity);

-			params->identities[params->nusers] = state->optarg;
+			params->identities[params->nusers] = optarg;
 			params->nusers++;
 			if (params->eap_id_length > WIFI_ENT_IDENTITY_MAX_LEN) {
 				PR_WARNING("eap identity too long (max %d characters)\n",
@@ -910,10 +910,10 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 				return -EINVAL;
 			}

-			params->eap_password = state->optarg;
+			params->eap_password = optarg;
 			params->eap_passwd_length = strlen(params->eap_password);

-			params->passwords[params->passwds] = state->optarg;
+			params->passwords[params->passwds] = optarg;
 			params->passwds++;
 			if (params->eap_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 				PR_WARNING("eap password length too long (max %d characters)\n",
@@ -925,7 +925,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 			params->ft_used = true;
 			break;
 		case 'g':
-			params->ignore_broadcast_ssid = shell_strtol(state->optarg, 10, &ret);
+			params->ignore_broadcast_ssid = shell_strtol(optarg, 10, &ret);
 			break;
 		case 'i':
 			/* Unused, but parsing to avoid unknown option error */
@@ -944,7 +944,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 			shell_help(sh);
 			return -ENOEXEC;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			return -EINVAL;
 		}

@@ -1126,7 +1126,7 @@ static int wifi_scan_args_to_params(const struct shell *sh,
 			opt_num++;
 			break;
 		case 'a':
-			val = atoi(state->optarg);
+			val = atoi(optarg);

 			if (val < 0) {
 				PR_ERROR("Invalid dwell_time_active val\n");
@@ -1137,7 +1137,7 @@ static int wifi_scan_args_to_params(const struct shell *sh,
 			opt_num++;
 			break;
 		case 'p':
-			val = atoi(state->optarg);
+			val = atoi(optarg);

 			if (val < 0) {
 				PR_ERROR("Invalid dwell_time_passive val\n");
@@ -1148,7 +1148,7 @@ static int wifi_scan_args_to_params(const struct shell *sh,
 			opt_num++;
 			break;
 		case 's':
-			if (wifi_utils_parse_scan_ssids(state->optarg,
+			if (wifi_utils_parse_scan_ssids(optarg,
 							params->ssids,
 							ARRAY_SIZE(params->ssids))) {
 				PR_ERROR("Invalid SSID(s)\n");
@@ -1158,7 +1158,7 @@ static int wifi_scan_args_to_params(const struct shell *sh,
 			opt_num++;
 			break;
 		case 'm':
-			val = atoi(state->optarg);
+			val = atoi(optarg);

 			if ((val < 0) || (val > WIFI_MGMT_SCAN_MAX_BSS_CNT)) {
 				PR_ERROR("Invalid max_bss val\n");
@@ -1169,7 +1169,7 @@ static int wifi_scan_args_to_params(const struct shell *sh,
 			opt_num++;
 			break;
 		case 'c':
-			if (wifi_utils_parse_scan_chan(state->optarg,
+			if (wifi_utils_parse_scan_chan(optarg,
 						       params->band_chan,
 						       ARRAY_SIZE(params->band_chan))) {
 				PR_ERROR("Invalid band or channel value(s)\n");
@@ -1837,7 +1837,7 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 'c':
-			if (!parse_number(sh, &value, state->optarg, NULL,
+			if (!parse_number(sh, &value, optarg, NULL,
 					  WIFI_TWT_SETUP_CMD_REQUEST,
 					  WIFI_TWT_SETUP_CMD_DEMAND)) {
 				return -EINVAL;
@@ -1846,14 +1846,14 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 't':
-			if (!parse_number(sh, &value, state->optarg, NULL, 1, 255)) {
+			if (!parse_number(sh, &value, optarg, NULL, 1, 255)) {
 				return -EINVAL;
 			}
 			params->dialog_token = (uint8_t)value;
 			break;

 		case 'f':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			if (!parse_number(sh, &value, optarg, NULL, 0,
 					  (WIFI_MAX_TWT_FLOWS - 1))) {
 				return -EINVAL;
 			}
@@ -1861,35 +1861,35 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 'r':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.responder = (bool)value;
 			break;

 		case 'T':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.trigger = (bool)value;
 			break;

 		case 'I':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.implicit = (bool)value;
 			break;

 		case 'a':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.announce = (bool)value;
 			break;

 		case 'w':
-			if (!parse_number(sh, &value, state->optarg, NULL, 1,
+			if (!parse_number(sh, &value, optarg, NULL, 1,
 					  WIFI_MAX_TWT_WAKE_INTERVAL_US)) {
 				return -EINVAL;
 			}
@@ -1897,7 +1897,7 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 'p':
-			if (!parse_number(sh, &value, state->optarg, NULL, 1,
+			if (!parse_number(sh, &value, optarg, NULL, 1,
 					  WIFI_MAX_TWT_INTERVAL_US)) {
 				return -EINVAL;
 			}
@@ -1905,7 +1905,7 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 'D':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			if (!parse_number(sh, &value, optarg, NULL, 0,
 					  WIFI_MAX_TWT_WAKE_AHEAD_DURATION_US)) {
 				return -EINVAL;
 			}
@@ -1913,14 +1913,14 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 'd':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 1)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 1)) {
 				return -EINVAL;
 			}
 			params->setup.twt_info_disable = (bool)value;
 			break;

 		case 'e':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			if (!parse_number(sh, &value, optarg, NULL, 0,
 					  WIFI_MAX_TWT_EXPONENT)) {
 				return -EINVAL;
 			}
@@ -1928,7 +1928,7 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 			break;

 		case 'm':
-			if (!parse_number(sh, &value, state->optarg, NULL, 0, 0xFFFF)) {
+			if (!parse_number(sh, &value, optarg, NULL, 0, 0xFFFF)) {
 				return -EINVAL;
 			}
 			params->setup.twt_mantissa = (uint16_t)value;
@@ -2231,7 +2231,7 @@ static int wifi_ap_config_args_to_params(const struct shell *sh, size_t argc, ch
 			params->type |= WIFI_AP_CONFIG_PARAM_MAX_INACTIVITY;
 			break;
 		case 's':
-			if (!parse_number(sh, &val, state->optarg, "max_num_sta",
+			if (!parse_number(sh, &val, optarg, "max_num_sta",
 					  0, CONFIG_WIFI_MGMT_AP_MAX_NUM_STA)) {
 				return -EINVAL;
 			}
@@ -2240,11 +2240,11 @@ static int wifi_ap_config_args_to_params(const struct shell *sh, size_t argc, ch
 			break;
 #if defined(CONFIG_WIFI_NM_HOSTAPD_AP)
 		case 'n':
-			strncpy(params->ht_capab, state->optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
+			strncpy(params->ht_capab, optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
 			params->type |= WIFI_AP_CONFIG_PARAM_HT_CAPAB;
 			break;
 		case 'c':
-			strncpy(params->vht_capab, state->optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
+			strncpy(params->vht_capab, optarg, WIFI_AP_IEEE_80211_CAPAB_MAX_LEN);
 			params->type |= WIFI_AP_CONFIG_PARAM_VHT_CAPAB;
 			break;
 #endif
@@ -2255,7 +2255,7 @@ static int wifi_ap_config_args_to_params(const struct shell *sh, size_t argc, ch
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		}
@@ -2734,7 +2734,7 @@ void parse_mode_args_to_params(const struct shell *sh, int argc,
 			opt_num++;
 			break;
 		case 'i':
-			mode->if_index = (uint8_t)atoi(state->optarg);
+			mode->if_index = (uint8_t)atoi(optarg);
 			/* Don't count iface as it's common for both get and set */
 			break;
 		case 'h':
@@ -2971,10 +2971,10 @@ void parse_filter_args_to_params(const struct shell *sh, int argc,
 			filter->filter |= WIFI_PACKET_FILTER_DATA;
 			break;
 		case 'i':
-			filter->if_index = (uint8_t)atoi(state->optarg);
+			filter->if_index = (uint8_t)atoi(optarg);
 			break;
 		case 'b':
-			filter->buffer_size = (uint16_t)atoi(state->optarg);
+			filter->buffer_size = (uint16_t)atoi(optarg);
 			break;
 		case 'h':
 			shell_help(sh);
@@ -4298,7 +4298,7 @@ static int wifi_bgscan_args_to_params(const struct shell *sh, size_t argc, char
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		}
@@ -4361,7 +4361,7 @@ static int wifi_config_args_to_params(const struct shell *sh, size_t argc, char
 			/* Unused, but parsing to avoid unknown option error */
 			break;
 		default:
-			PR_ERROR("Invalid option %c\n", state->optopt);
+			PR_ERROR("Invalid option %c\n", optopt);
 			shell_help(sh);
 			return SHELL_CMD_HELP_PRINTED;
 		}
diff --git a/subsys/net/lib/wifi_credentials/Kconfig b/subsys/net/lib/wifi_credentials/Kconfig
index 1a5671ccd73..0d53c2b1136 100644
--- a/subsys/net/lib/wifi_credentials/Kconfig
+++ b/subsys/net/lib/wifi_credentials/Kconfig
@@ -57,8 +57,6 @@ config WIFI_CREDENTIALS_SHELL
 	bool "Shell commands to manage Wi-Fi credentials"
 	default y
 	depends on SHELL
-	select SHELL_GETOPT
-	select GETOPT_LONG
 	depends on !WIFI_CREDENTIALS_BACKEND_NONE

 config WIFI_CREDENTIALS_CONNECT_STORED
diff --git a/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c b/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
index 7f2402120c9..27b4bf082b6 100644
--- a/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
+++ b/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
@@ -174,12 +174,12 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 					   WIFI_PSK_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.password, state->optarg, creds.password_len);
+			memcpy(creds.password, optarg, creds.password_len);
 			break;
 		case 'c':
-			channel = strtol(state->optarg, &endptr, 10);
+			channel = strtol(optarg, &endptr, 10);
 			if (*endptr != '\0') {
-				shell_error(sh, "Invalid channel: %s\n", state->optarg);
+				shell_error(sh, "Invalid channel: %s\n", optarg);
 				return -EINVAL;
 			}

@@ -210,7 +210,7 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 			creds.header.channel = channel;
 			break;
 		case 'b':
-			switch (atoi(state->optarg)) {
+			switch (atoi(optarg)) {
 			case 2:
 				creds.header.flags |= WIFI_CREDENTIALS_FLAG_2_4GHz;
 				break;
@@ -221,7 +221,7 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 				creds.header.flags |= WIFI_CREDENTIALS_FLAG_6GHz;
 				break;
 			default:
-				shell_error(sh, "Invalid band: %d\n", atoi(state->optarg));
+				shell_error(sh, "Invalid band: %d\n", atoi(optarg));
 				return -EINVAL;
 			}
 			break;
@@ -232,9 +232,9 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 					    wifi_security_txt(creds.header.type));
 				return -ENOTSUP;
 			}
-			mfp = strtol(state->optarg, &endptr, 10);
+			mfp = strtol(optarg, &endptr, 10);
 			if (*endptr != '\0') {
-				shell_error(sh, "Invalid IEEE 802.11w value: %s", state->optarg);
+				shell_error(sh, "Invalid IEEE 802.11w value: %s", optarg);
 				return -EINVAL;
 			}
 			if (mfp == WIFI_MFP_DISABLE) {
@@ -242,45 +242,43 @@ static int cmd_add_network(const struct shell *sh, size_t argc, char *argv[])
 			} else if (mfp == WIFI_MFP_REQUIRED) {
 				creds.header.flags |= WIFI_CREDENTIALS_FLAG_MFP_REQUIRED;
 			} else if (mfp > 2) {
-				shell_error(sh, "Invalid IEEE 802.11w value: %s",
-					    state->optarg);
+				shell_error(sh, "Invalid IEEE 802.11w value: %s", optarg);
 				return -EINVAL;
 			}
 			break;
 		case 'm':
 			if (net_bytes_from_str(creds.header.bssid, sizeof(creds.header.bssid),
-					       state->optarg) < 0) {
+					       optarg) < 0) {
 				shell_warn(sh, "Invalid MAC address\n");
 				return -EINVAL;
 			}
 			creds.header.flags |= WIFI_CREDENTIALS_FLAG_BSSID;
 			break;
 		case 'a':
-			creds.header.aid_length = strlen(state->optarg);
+			creds.header.aid_length = strlen(optarg);
 			if (creds.header.aid_length > WIFI_ENT_IDENTITY_MAX_LEN) {
 				shell_warn(sh, "anon_id too long (max %d characters)\n",
 					   WIFI_ENT_IDENTITY_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.header.anon_id, state->optarg, creds.header.aid_length);
+			memcpy(creds.header.anon_id, optarg, creds.header.aid_length);
 			creds.header.flags |= WIFI_CREDENTIALS_FLAG_ANONYMOUS_IDENTITY;
 			break;
 		case 'K':
-			creds.header.key_passwd_length = strlen(state->optarg);
+			creds.header.key_passwd_length = strlen(optarg);
 			if (creds.header.key_passwd_length > WIFI_ENT_PSWD_MAX_LEN) {
 				shell_warn(sh, "key_passwd too long (max %d characters)\n",
 					   WIFI_ENT_PSWD_MAX_LEN);
 				return -EINVAL;
 			}
-			memcpy(creds.header.key_passwd, state->optarg,
-			       creds.header.key_passwd_length);
+			memcpy(creds.header.key_passwd, optarg, creds.header.key_passwd_length);
 			creds.header.flags |= WIFI_CREDENTIALS_FLAG_KEY_PASSWORD;
 			break;
 		case 'h':
 			shell_help(sh);
 			return -ENOEXEC;
 		default:
-			shell_error(sh, "Invalid option %c\n", state->optopt);
+			shell_error(sh, "Invalid option %c\n", optopt);
 			return -EINVAL;
 		}
 	}
diff --git a/subsys/shell/Kconfig b/subsys/shell/Kconfig
index 4bbc1aef709..ce13fb8c4f8 100644
--- a/subsys/shell/Kconfig
+++ b/subsys/shell/Kconfig
@@ -214,7 +214,6 @@ config SHELL_HELP
 config SHELL_HELP_OPT_PARSE
 	bool "Parse -h and --help options"
 	depends on SHELL_HELP
-	depends on !SHELL_GETOPT
 	default y
 	help
 	  Shell parses command to find '-h' or '--help' string. If the shell
diff --git a/subsys/testsuite/ztest/Kconfig b/subsys/testsuite/ztest/Kconfig
index 6cae40e2869..131423ee2c5 100644
--- a/subsys/testsuite/ztest/Kconfig
+++ b/subsys/testsuite/ztest/Kconfig
@@ -27,8 +27,6 @@ config ZTEST_SHELL
 	bool "Ztest with shell support"
 	select SHELL
 	select SHELL_THREAD_PRIORITY_OVERRIDE
-	select GETOPT_LONG
-	select SHELL_GETOPT
 	help
 	  Enable shell to manage test execution and selection.

diff --git a/subsys/testsuite/ztest/src/ztest.c b/subsys/testsuite/ztest/src/ztest.c
index aeefbf7b494..d5599e0fd60 100644
--- a/subsys/testsuite/ztest/src/ztest.c
+++ b/subsys/testsuite/ztest/src/ztest.c
@@ -1278,7 +1278,7 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
 			opt_num++;
 			break;
 		case 'c':
-			val = atoi(state->optarg);
+			val = atoi(optarg);
 			if (val < 1) {
 				shell_error(sh, "Invalid number of case iterations");
 				return -ENOEXEC;
@@ -1326,7 +1326,7 @@ static int cmd_run_suite(const struct shell *sh, size_t argc, char **argv)
 			opt_num++;
 			break;
 		case 'p':
-			param = state->optarg;
+			param = optarg;
 			opt_num++;
 			break;
 		default:
