diff --git a/modules/hostap/Kconfig b/modules/hostap/Kconfig
index 354eacaef98..cff171e02f1 100644
--- a/modules/hostap/Kconfig
+++ b/modules/hostap/Kconfig
@@ -9,7 +9,13 @@ config WIFI_NM_WPA_SUPPLICANT
 	bool "WPA Supplicant from hostap project [EXPERIMENTAL]"
 	select POSIX_TIMERS
 	select POSIX_SIGNALS
-	select POSIX_API
+	select POSIX_SYSTEM_INTERFACES
+	select POSIX_BASE_DEFINITIONS
+	select POSIX_AEP_REALTIME_MINIMAL
+	select POSIX_DEVICE_IO
+	select POSIX_FD_MGMT
+	select POSIX_MULTI_PROCESS
+	select XSI
 	select XSI_SINGLE_PROCESS
 	select NET_SOCKETS
 	select NET_SOCKETS_PACKET
