diff --git a/include/zephyr/kernel.h b/include/zephyr/kernel.h
index 46ecc0ab1de..193c811b9ac 100644
--- a/include/zephyr/kernel.h
+++ b/include/zephyr/kernel.h
@@ -7106,6 +7106,127 @@ __syscall bool k_thread_cleanup_pop(int execute);
 
 #endif /* CONFIG_THREAD_CLEANUP */
 
+/**
+ * @}
+ */
+
+/**
+ * @addtogroup k_thread_specific_storage_apis
+ * @{
+ */
+
+/**
+ * @brief Destructor type for the thread-specific storage values associated with a given key.
+ *
+ * @param value Value to destroy.
+ */
+typedef void (*k_thread_key_destructor_t)(void *value);
+
+#if defined(CONFIG_THREAD_SPECIFIC_STORAGE) || defined(__DOXYGEN__)
+/**
+ * @brief Wildcard key selector for thread-specific storage queries.
+ *
+ * Used with @ref k_thread_qspecific and @ref k_thread_key_delete to match any stored key.
+ */
+#define K_THREAD_KEY_ANY ((void *)(-1))
+
+/**
+ * @brief Create a thread-specific storage key.
+ *
+ * Create a thread-specific storage key and optionally associate a destructor with it.
+ *
+ * Destructors functions are only ever called during graceful thread termination (i.e. when
+ * a thread exits by running to completion, via @ref k_thread_cancel, or via @ref k_thread_exit).
+ *
+ * @param[out] key Pointer to location where the created key is stored.
+ * @param dtor Optional destructor for values associated with this key.
+ *
+ * @retval 0 on success.
+ * @retval -ENOMEM if no keys are available.
+ */
+__syscall int k_thread_key_create(void **key, k_thread_key_destructor_t dtor);
+
+/**
+ * @brief Delete a thread-specific storage key.
+ *
+ * @param key Key to delete.
+ *
+ * @retval 0 on success.
+ * @retval -EINVAL if @p key is invalid.
+ * @retval -EBUSY if @p key has active thread-specific associations.
+ * @retval -ENOENT if @p key is @ref K_THREAD_KEY_ANY and no thread-specific values are stored.
+ */
+__syscall int k_thread_key_delete(void *key);
+
+/**
+ * @brief Set a thread-specific value for the calling thread.
+ *
+ * @param key Key identifying the thread-specific value.
+ * @param value Value to associate with @p key.
+ *
+ * @retval 0 on success.
+ * @retval -EINVAL if @p key is invalid.
+ * @retval -ENOMEM if memory for a new key/value association cannot be allocated.
+ */
+__syscall int k_thread_setspecific(void *key, void *value);
+
+/**
+ * @brief Query a thread-specific value from the calling thread.
+ *
+ * If @p key points to a valid key for which there is an associated thread-specific value, then
+ * the value is written to the memory location pointed to by @p value. The key-value pair remains
+ * associated for the calling thread.
+ *
+ * If @p key points to @ref K_THREAD_KEY_ANY, then any available entry is selected and the
+ * associated key-value pair is removed from thread-specific storage for the calling thread.
+ *
+ * @param[in,out] key On input, Pointer to the key selector. On success, the key associated with @p
+ * value.
+ * @param[out] value Pointer to where the value associated with @p key is stored.
+ * @param[out] dtor Pointer to where the destructor associated with the removed value is stored.
+ *
+ * @retval 0 on success.
+ * @retval -EINVAL if @p key is invalid, or if @p *key has no association in the calling thread.
+ * @retval -ENOENT if @p key is @ref K_THREAD_KEY_ANY and the calling thread has no stored values.
+ */
+__syscall int k_thread_qspecific(void **key, void **value, k_thread_key_destructor_t *dtor);
+
+/**
+ * @brief Remove and return any thread-specific value from the calling thread.
+ *
+ * Convenience wrapper for k_thread_qspecific() with @ref K_THREAD_KEY_ANY.
+ *
+ * @param[in,out] key Pointer to where the key associated with @p value is stored.
+ * @param[out] value Pointer to where the value associated with @p key is stored.
+ * @param[out] dtor Pointer to where the destructor associated with the removed value is stored.
+ *
+ * @retval 0 on success.
+ * @retval -ENOENT if the calling thread has no stored values.
+ */
+static inline int k_thread_popspecific(void **key, void **value, k_thread_key_destructor_t *dtor)
+{
+	*key = K_THREAD_KEY_ANY;
+	return k_thread_qspecific(key, value, dtor);
+}
+
+/**
+ * @brief Get a thread-specific value for a specific key.
+ *
+ * @param key Key identifying the thread-specific value.
+ * @param[out] value Pointer to where the value associated with @p key is stored.
+ *
+ * @retval 0 on success.
+ * @retval -EINVAL if @p key is invalid, or if @p key has no association in the calling thread.
+ */
+static inline int k_thread_getspecific(void *key, void **value)
+{
+	__maybe_unused k_thread_key_destructor_t dtor;
+
+	return k_thread_qspecific(&key, value, &dtor);
+}
+
+#endif /* CONFIG_THREAD_SPECIFIC_STORAGE */
+
 /**
  * @}
  */
diff --git a/include/zephyr/kernel/thread.h b/include/zephyr/kernel/thread.h
index 939052af2ab..da690b21b6e 100644
--- a/include/zephyr/kernel/thread.h
+++ b/include/zephyr/kernel/thread.h
@@ -153,6 +153,11 @@ struct _thread_base {
        /** Cleanup handlers */
        sys_slist_t cleanup;
 #endif
+
+#ifdef CONFIG_THREAD_SPECIFIC_STORAGE
+       /** Thread-specific storage keys */
+       sys_dlist_t key_list;
+#endif
 };
 
 typedef struct _thread_base _thread_base_t;
diff --git a/kernel/CMakeLists.txt b/kernel/CMakeLists.txt
index 75e71f5033f..440c5ae7a1c 100644
--- a/kernel/CMakeLists.txt
+++ b/kernel/CMakeLists.txt
@@ -3,6 +3,7 @@
 # This file cannot be compiled with __ZEPHYR_SUPERVISOR__ when userspace is enabled
 zephyr_sources_ifdef(CONFIG_SIGNAL signal.c)
 zephyr_sources_ifdef(CONFIG_THREAD_CLEANUP cleanup.c)
+zephyr_sources_ifdef(CONFIG_THREAD_SPECIFIC_STORAGE tss.c)
 
 # kernel is a normal CMake library and not a zephyr_library because it
 # should usually not be --whole-archive'd
diff --git a/kernel/Kconfig b/kernel/Kconfig
index a7a0112ec90..cdc0ca8163e 100644
--- a/kernel/Kconfig
+++ b/kernel/Kconfig
@@ -220,6 +220,26 @@ config THREAD_RESULT
 	help
 	  Say 'Y' here to support a thread result or "return value".
 
+config THREAD_SPECIFIC_STORAGE
+	bool "Thread specific storage (TSS)"
+	help
+	  Say 'Y' here to support thread-specific storage. That feature allows the
+	  application to add and remove thread-specific values at runtime, in contrast to
+	  thread-local storage (TLS) which requires values be annotated `__thread` at
+	  compile time. TSS can also be used to emulated TLS on systems which lack hardware TLS
+	  support.
+
+if THREAD_SPECIFIC_STORAGE
+
+config THREAD_SPECIFIC_STORAGE_KEYS_MAX
+	int "Maximum number of thread-specific storage keys"
+	default 1
+	help
+	  Adjust this number based on the maximum number of thread-specific storage keys
+	  that will be needed for the application.
+
+endif
+
 config THREAD_CANCEL_SIGNAL_NUMBER
 	int "Thread cancellation signal number"
 	def_int 32
diff --git a/kernel/signal.c b/kernel/signal.c
index 59f1a57bdf1..672c27538db 100644
--- a/kernel/signal.c
+++ b/kernel/signal.c
@@ -866,7 +866,18 @@ static void z_sig_handler_cancel(int sig, struct k_sig_info *info, void *ucontex
 	}
 #endif
 
-	/* destroy thread-specific storage (not yet implemented) */
+#ifdef CONFIG_THREAD_SPECIFIC_STORAGE
+	/* destroy thread-specific storage */
+	for (void *key = NULL, *value = NULL, *_dtor = NULL;
+	     k_thread_popspecific(&key, &value, (k_thread_key_destructor_t *)&_dtor) == 0;) {
+		k_thread_key_destructor_t dtor = (k_thread_key_destructor_t)_dtor;
+
+		if (dtor != NULL) {
+			dtor(value);
+		}
+	}
+#endif
+
 	z_sig_handler_term(sig, info, ucontext);
 }
 #endif /* CONFIG_THREAD_CANCEL */
diff --git a/kernel/thread.c b/kernel/thread.c
index 6a17a469134..b71b3f02efd 100644
--- a/kernel/thread.c
+++ b/kernel/thread.c
@@ -741,6 +741,10 @@ char *z_setup_new_thread(struct k_thread *new_thread,
 	sys_slist_init(&new_thread->base.cleanup);
 #endif
 
+#ifdef CONFIG_THREAD_SPECIFIC_STORAGE
+	sys_dlist_init(&new_thread->base.key_list);
+#endif
+
 	SYS_PORT_TRACING_OBJ_FUNC(k_thread, create, new_thread);
 
 	return stack_ptr;
diff --git a/kernel/tss.c b/kernel/tss.c
new file mode 100644
index 00000000000..265e0e3f5da
--- /dev/null
+++ b/kernel/tss.c
@@ -0,0 +1,380 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2023 Meta
+ * Copyright (c) 2025, Friedt Professional Engineering Services, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/internal/syscall_handler.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+
+#if (CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX % BITS_PER_LONG) == 0
+#define K_TSS_LAST_WORD_MASK (~0UL)
+#else
+#ifdef CONFIG_64BIT
+#define K_TSS_LAST_WORD_MASK BIT64_MASK(CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX % BITS_PER_LONG)
+#else
+#define K_TSS_LAST_WORD_MASK BIT_MASK(CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX % BITS_PER_LONG)
+#endif
+#endif
+
+struct k_thread_key_obj {
+	/* List of k_thread_key_data objects that contain thread specific data for the key */
+	sys_dlist_t key_data_l;
+
+	/* Optional destructor that is passed to k_thread_key_create() */
+	void (*destructor)(void *value);
+};
+
+struct k_thread_thread_data {
+	sys_dnode_t node;
+
+	/* Key and thread specific data passed to k_thread_setspecific() */
+	void *val;
+	void *key;
+};
+
+struct k_thread_key_data {
+	sys_dnode_t node;
+	struct k_thread_thread_data thread_data;
+};
+
+LOG_MODULE_DECLARE(os, CONFIG_KERNEL_LOG_LEVEL);
+
+static struct k_spinlock lock;
+__noinit static struct k_thread_key_obj pool[CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX];
+static unsigned long
+	allocation_bitmap[DIV_ROUND_UP(CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX, BITS_PER_LONG)];
+
+/* like find_lsb_set() but operates on longs rather than fixed-size uint32_t's */
+static ALWAYS_INLINE unsigned int find_lsb_set_l(unsigned long op)
+{
+#ifdef CONFIG_TOOLCHAIN_HAS_BUILTIN_FFS
+	return __builtin_ffsl(op);
+#else
+	if (op == 0) {
+		return 0;
+	}
+
+	for (size_t i = 0; i < sizeof(op) * BITS_PER_BYTE; ++i, op >>= 1) {
+		if ((op & 1) != 0) {
+			return i + 1;
+		}
+	}
+
+	return -1;
+#endif
+}
+
+static bool k_thread_key_is_allocated(size_t key_index)
+{
+	size_t word_index = key_index / BITS_PER_LONG;
+	size_t bit_index = key_index % BITS_PER_LONG;
+#ifdef CONFIG_64BIT
+	unsigned long mask = BIT64(bit_index);
+#else
+	unsigned long mask = BIT(bit_index);
+#endif
+
+	return (allocation_bitmap[word_index] & mask) != 0;
+}
+
+static bool k_thread_key_is_valid(void *key)
+{
+	return IS_ARRAY_ELEMENT(pool, key) &&
+	       k_thread_key_is_allocated((struct k_thread_key_obj *)key - pool);
+}
+
+/* find the least-significant unset bit in allocation_bitmap */
+static void *k_thread_key_alloc(void)
+{
+	int bit_index;
+	unsigned long word;
+	unsigned long mask;
+	struct k_thread_key_obj *key_obj = NULL;
+
+	ARRAY_FOR_EACH(allocation_bitmap, word_index) {
+		word = ~allocation_bitmap[word_index];
+		/* ones in word indicate indeces of available keys */
+		if (word_index == ARRAY_SIZE(allocation_bitmap) - 1) {
+			/* clear most-significant bits of the last word */
+			word &= K_TSS_LAST_WORD_MASK;
+		}
+
+		bit_index = find_lsb_set_l(word) - 1;
+		if (bit_index < 0) {
+			continue;
+		}
+
+		mask = IS_ENABLED(CONFIG_64BIT) ? BIT64(bit_index) : BIT(bit_index);
+		allocation_bitmap[word_index] |= mask;
+		key_obj = &pool[word_index * BITS_PER_LONG + bit_index];
+		break;
+	}
+
+	return key_obj;
+}
+
+static void k_thread_key_free(void *key)
+{
+	size_t key_index = (struct k_thread_key_obj *)key - pool;
+	size_t word_index = key_index / BITS_PER_LONG;
+	size_t bit_index = key_index % BITS_PER_LONG;
+	unsigned long mask = IS_ENABLED(CONFIG_64BIT) ? BIT(bit_index) : BIT(bit_index);
+
+	allocation_bitmap[word_index] &= ~mask;
+}
+
+int z_impl_k_thread_key_create(void **key, k_thread_key_destructor_t destructor)
+{
+	struct k_thread_key_obj *key_obj = NULL;
+
+	K_SPINLOCK(&lock) {
+		key_obj = k_thread_key_alloc();
+	}
+	if (key_obj == NULL) {
+		/* no objects left to allocate */
+		return -ENOMEM;
+	}
+
+	*key = key_obj;
+	sys_dlist_init(&key_obj->key_data_l);
+	key_obj->destructor = destructor;
+
+	LOG_DBG("Initialized key %p", key_obj);
+
+	return 0;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_thread_key_create(void **key, k_thread_key_destructor_t destructor)
+{
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(key, sizeof(*key)));
+
+	return z_impl_k_thread_key_create(key, destructor);
+}
+#include <syscalls/k_thread_key_create_mrsh.c>
+#endif
+
+int z_impl_k_thread_key_delete(void *key)
+{
+	int ret = -EINVAL;
+
+	K_SPINLOCK(&lock) {
+
+		if (key == K_THREAD_KEY_ANY) {
+			bool found = false;
+
+			ARRAY_FOR_EACH(allocation_bitmap, word_index) {
+				/* ones in word indicate indeces of allocated keys */
+				unsigned long word = allocation_bitmap[word_index];
+				int bit_index = find_lsb_set_l(word) - 1;
+
+				if (bit_index < 0) {
+					continue;
+				}
+
+				key = &pool[word_index * BITS_PER_LONG + bit_index];
+				found = true;
+				break;
+			}
+
+			if (!found) {
+				ret = -ENOENT;
+				K_SPINLOCK_BREAK;
+			}
+		} else {
+			if (!k_thread_key_is_valid(key)) {
+				LOG_DBG("Attempt to free invalid key %p", key);
+				ret = -EINVAL;
+				K_SPINLOCK_BREAK;
+			}
+
+			struct k_thread_key_obj *const key_obj = key;
+
+			if (sys_dlist_len(&(key_obj->key_data_l)) != 0) {
+				LOG_DBG("Attempt to delete key %p while associations exist", key);
+				ret = -EBUSY;
+				K_SPINLOCK_BREAK;
+			}
+		}
+
+		k_thread_key_free(key);
+		ret = 0;
+	}
+
+	if (ret == 0) {
+		LOG_DBG("Deleted key %p", key);
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_thread_key_delete(void *key)
+{
+	return z_impl_k_thread_key_delete(key);
+}
+#include <syscalls/k_thread_key_delete_mrsh.c>
+#endif
+
+int z_impl_k_thread_setspecific(void *key, void *value)
+{
+	int retval = 0;
+
+	K_SPINLOCK(&lock) {
+		struct k_thread_key_obj *const key_obj = key;
+		struct k_thread *const thread = k_current_get();
+		struct k_thread_key_data *key_data;
+		sys_dnode_t *node_l = NULL;
+
+		if (!k_thread_key_is_valid(key)) {
+			retval = -EINVAL;
+			K_SPINLOCK_BREAK;
+		}
+
+		SYS_DLIST_FOR_EACH_NODE(&thread->base.key_list, node_l) {
+			key_data = CONTAINER_OF(node_l, struct k_thread_key_data, node);
+
+			if (key_data->thread_data.key == key) {
+				/* Key is already present so associate new thread specific data
+				 *
+				 * Note: it is responsibility of the application to call any
+				 * destructor associated with the key, for the old value,
+				 * independently. The only time that destructors are guaranteed to
+				 * be called is during graceful termination (i.e. running to
+				 * completion, `k_thread_exit()`, or `k_thread_cancel()`) and those
+				 * must be run in user context.
+				 */
+				key_data->thread_data.val = value;
+				LOG_DBG("Paired key %p to value %p for thread %p", key, value,
+					k_current_get());
+				break;
+			}
+		}
+
+		retval = 0;
+		if (node_l != NULL) {
+			/* Key is already present, so we are done */
+			K_SPINLOCK_BREAK;
+		}
+
+#if (K_HEAP_MEM_POOL_SIZE > 0)
+		/* Key and data need to be added */
+		key_data = k_malloc(sizeof(struct k_thread_key_data));
+#else
+		key_data = NULL;
+#endif
+
+		if (key_data == NULL) {
+			LOG_DBG("Failed to allocate key data for key %p", key);
+			retval = -ENOMEM;
+			K_SPINLOCK_BREAK;
+		}
+
+		sys_dnode_init(&key_data->node);
+		sys_dnode_init(&key_data->thread_data.node);
+
+		/* Associate thread specific data, initialize new key */
+		key_data->thread_data.key = key;
+		key_data->thread_data.val = value;
+
+		/* Append key data to the thread's key list */
+		sys_dlist_append(&thread->base.key_list, &key_data->node);
+
+		/* Append thread data to the key object's list */
+		sys_dlist_append(&key_obj->key_data_l, &key_data->thread_data.node);
+		LOG_DBG("Created key-value pair (%p, %p)@%p for thread %p", key, value, key_data,
+			k_current_get());
+	}
+
+	return retval;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_thread_setspecific(void *key, void *value)
+{
+	return z_impl_k_thread_setspecific(key, value);
+}
+#include <syscalls/k_thread_setspecific_mrsh.c>
+#endif
+
+int z_impl_k_thread_qspecific(void **key, void **value, k_thread_key_destructor_t *dtor)
+{
+	int ret = -EINVAL;
+
+	K_SPINLOCK(&lock) {
+		struct k_thread_key_obj *key_obj = NULL;
+		struct k_thread *const tid = _current;
+		struct k_thread_key_data *key_data;
+		sys_dnode_t *node_l;
+
+		if (*key == K_THREAD_KEY_ANY) {
+			/* Remove key data from the thread's key list */
+			node_l = sys_dlist_get(&tid->base.key_list);
+			if (node_l == NULL) {
+				ret = -ENOENT;
+				K_SPINLOCK_BREAK;
+			}
+
+			key_data = CONTAINER_OF(node_l, struct k_thread_key_data, node);
+			*key = key_data->thread_data.key;
+			*value = key_data->thread_data.val;
+			key_obj = *key;
+			*dtor = key_obj->destructor;
+
+			/* Remove thread data from the key object's list */
+			sys_dlist_remove(&key_data->thread_data.node);
+
+#if (K_HEAP_MEM_POOL_SIZE > 0)
+			k_free(key_data);
+#endif
+
+			LOG_DBG("Removed key-value pair (%p, %p)@%p for thread %p", *key, *value,
+				key_data, k_current_get());
+
+			ret = 0;
+			K_SPINLOCK_BREAK;
+		}
+
+		if (!k_thread_key_is_valid(*key)) {
+			LOG_DBG("Invalid key %p for thread %p", *key, k_current_get());
+			ret = -EINVAL;
+			K_SPINLOCK_BREAK;
+		}
+
+		key_obj = *key;
+
+		/* Traverse the list of keys set by the thread, looking for key */
+
+		SYS_DLIST_FOR_EACH_NODE(&tid->base.key_list, node_l) {
+			key_data = CONTAINER_OF(node_l, struct k_thread_key_data, node);
+			if (key_data->thread_data.key == *key) {
+				LOG_DBG("Found key-value pair (%p, %p)@%p for thread %p", *key,
+					key_data->thread_data.val, key_data, k_current_get());
+				/* Key is present, so get the set thread data */
+				*value = key_data->thread_data.val;
+				*dtor = key_obj->destructor;
+				ret = 0;
+				break;
+			}
+		}
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_thread_qspecific(void **key, void **value, k_thread_key_destructor_t *dtor)
+{
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(key, sizeof(*key)));
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(value, sizeof(*value)));
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(dtor, sizeof(*dtor)));
+
+	return z_impl_k_thread_qspecific(key, value, dtor);
+}
+#include <syscalls/k_thread_qspecific_mrsh.c>
+#endif
diff --git a/tests/kernel/threads/thread_apis/prj.conf b/tests/kernel/threads/thread_apis/prj.conf
index bf76aff7bb5..a5e0c9bac3e 100644
--- a/tests/kernel/threads/thread_apis/prj.conf
+++ b/tests/kernel/threads/thread_apis/prj.conf
@@ -16,3 +16,6 @@ CONFIG_SIGNAL=y
 CONFIG_SIGNAL_SET_SIZE=32
 CONFIG_THREAD_CANCEL=y
 CONFIG_THREAD_CLEANUP=y
+CONFIG_HEAP_MEM_POOL_SIZE=256
+CONFIG_THREAD_SPECIFIC_STORAGE=y
+CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX=2
diff --git a/tests/kernel/threads/thread_apis/src/test_kthread_tss.c b/tests/kernel/threads/thread_apis/src/test_kthread_tss.c
new file mode 100644
index 00000000000..c283f83c968
--- /dev/null
+++ b/tests/kernel/threads/thread_apis/src/test_kthread_tss.c
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2026, Friedt Professional Engineering Services, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/sys/util.h>
+#include <zephyr/ztest.h>
+
+#define N_THR 2
+#define N_KEY 2
+
+ZTEST_BMEM static int entry_mode;
+ZTEST_BMEM static void *keys[N_KEY];
+ZTEST_BMEM static int alloc_count[N_THR];
+ZTEST_BMEM static int dtor_count;
+
+static void entry(void *arg1, void *arg2, void *arg3);
+
+/* array of thread stacks */
+#define K_THREAD_ARRAY_DEFINE_ELEM(idx, name_base, stack_size, entry, prio, options, delay)        \
+	K_THREAD_DEFINE(name_base##_##idx, stack_size, entry, NULL, NULL, NULL, prio, options,     \
+			delay)
+
+#define K_THREAD_ARRAY_ELEM_ID(idx, name_base) &_k_thread_obj_##name_base##_##idx
+
+#define K_THREAD_ARRAY_STACK_ELEM_ID(idx, name_base)                                               \
+	(k_thread_stack_t *)&_k_thread_stack_##name_base##_##idx
+
+#define K_THREAD_ARRAY_DEFINE_STATIC(name, num, stack_size, entry, prio, options, delay)           \
+	LISTIFY(num, K_THREAD_ARRAY_DEFINE_ELEM, (;), \
+		  name, stack_size, entry, prio, options, delay);      \
+	ZTEST_DMEM __maybe_unused static struct k_thread *name[] = {                               \
+		LISTIFY(num, K_THREAD_ARRAY_ELEM_ID, (,), name) };                                    \
+	ZTEST_DMEM __maybe_unused static k_thread_stack_t *name##_stack[] = {                      \
+		LISTIFY(num, K_THREAD_ARRAY_STACK_ELEM_ID, (,), name) }
+
+K_THREAD_ARRAY_DEFINE_STATIC(thread, N_THR, CONFIG_ZTEST_STACK_SIZE, entry, K_PRIO_PREEMPT(10),
+			     K_USER, K_TICKS_FOREVER);
+
+static int get_thread_index(void)
+{
+	for (int i = 0; i < N_THR; i++) {
+		if (k_current_get() == thread[i]) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+static void entry(void *arg1, void *arg2, void *arg3)
+{
+	void *value;
+	void *expected_value;
+
+	if (entry_mode == 0) {
+		value = (void *)0x42;
+		expected_value = value;
+		zassert_not_null(value, "thread could not allocate storage");
+		zassert_ok(k_thread_setspecific(keys[0], value), "k_thread_setspecific failed");
+		zassert_ok(k_thread_getspecific(keys[0], &value));
+		zassert_equal(value, expected_value, "set and retrieved values are different");
+	}
+
+	if (entry_mode == 1) {
+		value = (void *)0x73;
+		expected_value = value;
+		for (int i = 0; i < N_KEY; i++) {
+			zassert_ok(k_thread_setspecific(keys[i], value));
+			zassert_ok(k_thread_getspecific(keys[i], &value));
+			zassert_equal(value, expected_value);
+		}
+	}
+
+	if (entry_mode == 3) {
+		int value = 42;
+		int thread_index = get_thread_index();
+		int *const count = &alloc_count[thread_index];
+
+		while (1) {
+			BUILD_ASSERT((N_KEY % N_THR == 0) && (N_KEY / N_THR > 0),
+				     "N_KEY must be a multiple of N_THR for this test");
+			/*
+			 * This loop will keep creating and setting thread-specific keys until the
+			 * thread hits the limit on how many keys it can set. At that point, the
+			 * thread should have CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX / N_THR keys
+			 * set, and thus alloc_count should be equal across all threads, which is
+			 * verified in the test function. */
+			if (*count == N_KEY / N_THR) {
+				break;
+			}
+
+			if (k_thread_key_create(&keys[thread_index], NULL) < 0) {
+				break;
+			}
+			if (k_thread_setspecific(keys[thread_index], &value) == -ENOMEM) {
+				break;
+			};
+			*count += 1;
+		}
+	}
+
+	if (entry_mode == 4) {
+		for (int i = 0; i < N_KEY; i++) {
+			zassert_ok(k_thread_setspecific(keys[i], INT_TO_POINTER(0x42 + i)));
+		}
+	}
+}
+
+static void make_key(void)
+{
+	zassert_ok(k_thread_key_create(&keys[0], NULL), "insufficient memory to create key");
+}
+
+static void make_keys(void)
+{
+	for (int i = 0; i < N_KEY; i++) {
+		zassert_ok(k_thread_key_create(&keys[i], NULL),
+			   "insufficient memory to create keys");
+	}
+}
+
+ZTEST_USER(threads_tss, test_key_popspecific)
+{
+	void *value = NULL;
+	k_thread_key_destructor_t dtor = (k_thread_key_destructor_t)0x73;
+
+	zassert_equal(k_thread_popspecific(&keys[0], &value, &dtor), -ENOENT);
+	zassert_ok(k_thread_key_create(&keys[0], NULL));
+	zassert_ok(k_thread_setspecific(keys[0], (void *)0x42));
+	zassert_ok(k_thread_popspecific(&keys[1], &value, &dtor));
+	zassert_equal(keys[0], keys[1]);
+	zassert_equal(value, (void *)0x42);
+	zassert_equal(dtor, NULL);
+	zassert_ok(k_thread_key_delete(keys[0]));
+}
+
+ZTEST_USER(threads_tss, test_key_1toN_thread)
+{
+	make_key();
+
+	/* Different threads set different values to same key */
+	entry_mode = 0;
+	for (int i = 0; i < N_THR; i++) {
+		k_thread_start(thread[i]);
+	}
+
+	for (int i = 0; i < N_THR; i++) {
+		zassert_ok(k_thread_rejoin(thread[i], K_FOREVER, NULL), "failed to join thread %d",
+			   i);
+	}
+
+	/* At this point, there should be no more thread-specific values associated with key[0] */
+	zassert_ok(k_thread_key_delete(keys[0]), "attempt to delete key failed");
+}
+
+ZTEST_USER(threads_tss, test_key_Nto1_thread)
+{
+	make_keys();
+
+	/* Single thread associates its value with different keys */
+
+	entry_mode = 1;
+	k_thread_start(thread[0]);
+	zassert_ok(k_thread_rejoin(thread[0], K_FOREVER, NULL), "failed to join thread");
+
+	/* At this point, there should be no more thread-specific values associated with any key */
+	for (int i = 0; i < N_KEY; i++) {
+		zassert_ok(k_thread_key_delete(keys[i]), "attempt to delete keys[%d] failed", i);
+	}
+}
+
+ZTEST_USER(threads_tss, test_key_resource_leak)
+{
+	void *key;
+
+	for (size_t i = 0; i < 2 * CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX; ++i) {
+		zassert_ok(k_thread_key_create(&key, NULL), "failed to create key %zu", i);
+		zassert_ok(k_thread_key_delete(key), "failed to delete key %zu", i);
+	}
+}
+
+ZTEST_USER(threads_tss, test_correct_key_is_deleted)
+{
+	size_t j = CONFIG_THREAD_SPECIFIC_STORAGE_KEYS_MAX - 1;
+
+	for (size_t i = 0; i < ARRAY_SIZE(keys); ++i) {
+		zassert_ok(k_thread_key_create(&keys[i], NULL), "failed to create key %zu", i);
+	}
+
+	void *key = keys[j];
+	zassert_ok(k_thread_key_delete(keys[j]));
+	zassert_ok(k_thread_key_create(&keys[j], NULL), "failed to create key %zu", j);
+
+	zassert_equal(key, keys[j], "deleted key %p instead of key %p", keys[j], key);
+
+	for (size_t i = 0; i < ARRAY_SIZE(keys); ++i) {
+		zassert_ok(k_thread_key_delete(keys[i]), "failed to delete key %zu", i);
+	}
+}
+
+ZTEST_USER(threads_tss, test_thread_specific_data_allocation)
+{
+	entry_mode = 3;
+
+	for (int i = 0; i < N_THR; ++i) {
+		k_thread_start(thread[i]);
+		zassert_ok(k_thread_rejoin(thread[i], K_FOREVER, NULL));
+	}
+
+	zassert_equal(alloc_count[0], N_KEY / N_THR);
+
+	for (int i = 1; i < N_THR; ++i) {
+		zassert_equal(alloc_count[i - 1], alloc_count[i],
+			      "uneven distribution of thread keys. alloc_count[%d]: %d != "
+			      "alloc_count[%d]: %d",
+			      i, alloc_count[i], i - 1, alloc_count[i - 1]);
+	}
+
+	for (int i = 0; i < N_KEY; i++) {
+		zassert_ok(k_thread_key_delete(keys[i]));
+	}
+}
+
+static void tss_dtor(void *value)
+{
+	TC_PRINT("Thread %p: %s(%p)\n", k_current_get(), __func__, value);
+	++dtor_count;
+}
+
+static void make_keys_with_dtor(void)
+{
+	for (int i = 0; i < N_KEY; i++) {
+		zassert_ok(k_thread_key_create(&keys[i], tss_dtor));
+	}
+}
+
+ZTEST_USER(threads_tss, test_thread_specific_data_destructor)
+{
+	/*
+	 * In this test, we set thread-specific values in the child thread, but intentionally do not
+	 * delete them, in order to verify that they are destroyed via graceful thread termination
+	 * with k_thread_exit().
+	 */
+	make_keys_with_dtor();
+
+	entry_mode = 4;
+
+	for (int i = 0; i < N_THR; ++i) {
+		k_thread_start(thread[i]);
+		zassert_ok(k_thread_rejoin(thread[i], K_FOREVER, NULL));
+	}
+
+	zassert_equal(dtor_count, N_THR * N_KEY,
+		      "tss destructor call mismatch. expected: %d, actual: %d", N_THR * N_KEY,
+		      dtor_count);
+
+	for (int i = 0; i < N_KEY; i++) {
+		zassert_ok(k_thread_key_delete(keys[i]));
+	}
+}
+
+void threads_tss_before(void *arg)
+{
+	ARG_UNUSED(arg);
+
+#ifdef CONFIG_USERSPACE
+	for (int i = 0; i < N_THR; ++i) {
+		/* grant ztest_thread permissions on thread[i] */
+		k_thread_access_grant(k_current_get(), thread[i]);
+
+		/* So statically-defined userspace threads can be restarted with k_thread_start() */
+		k_object_init(thread[i]);
+		k_thread_create(thread[i], thread_stack[i], CONFIG_ZTEST_STACK_SIZE, entry, NULL,
+				NULL, NULL, K_PRIO_PREEMPT(10), K_USER, K_FOREVER);
+	}
+#endif
+}
+
+ZTEST_SUITE(threads_tss, NULL, NULL, threads_tss_before, NULL, NULL);
