diff --git a/kernel/sched.c b/kernel/sched.c
index f579928390f..33884f2bfdd 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -727,8 +727,12 @@ struct k_thread *z_unpend1_no_timeout(_wait_q_t *wait_q)

 void z_unpend_thread(struct k_thread *thread)
 {
-	z_unpend_thread_no_timeout(thread);
-	z_abort_thread_timeout(thread);
+	K_SPINLOCK(&_sched_spinlock) {
+		if (thread->base.pended_on != NULL) {
+			unpend_thread_no_timeout(thread);
+		}
+		z_abort_thread_timeout(thread);
+	}
 }

 /* Priority set utility that does no rescheduling, it just changes the
