diff --git a/kernel/userspace.c b/kernel/userspace.c
index 7e2304cec28..a68d0a74198 100644
--- a/kernel/userspace.c
+++ b/kernel/userspace.c
@@ -360,8 +360,9 @@ static struct k_object *dynamic_object_create(enum k_objects otype, size_t align
 		struct z_stack_data *stack_data = (struct z_stack_data *)
 			((uint8_t *)dyn->data + adjusted_size - sizeof(*stack_data));
 #if defined(CONFIG_ARM_MPU) || defined(CONFIG_ARC_MPU) || defined(CONFIG_RISCV_PMP)
-		stack_data->priv = (void *)ROUND_UP(((uint8_t *)dyn->data + size),
-			  Z_THREAD_STACK_OBJ_ALIGN(size));
+		stack_data->priv = (void *)ROUND_UP(
+			((uint8_t *)dyn->data + K_THREAD_STACK_LEN(size)),
+			Z_THREAD_STACK_OBJ_ALIGN(size));
 #else
 		stack_data->priv = (uint8_t *)dyn->data;
 #endif /* CONFIG_ARM_MPU || CONFIG_ARC_MPU || CONFIG_RISCV_PMP */
