diff --git a/kernel/init.c b/kernel/init.c
--- a/kernel/init.c
+++ b/kernel/init.c
@@ -328,6 +328,12 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
 	z_mem_manage_boot_finish();
 #endif /* CONFIG_MMU */
 
+#ifdef CONFIG_CURRENT_THREAD_USE_TLS
+	extern Z_THREAD_LOCAL k_tid_t z_tls_current;
+
+	z_tls_current = &z_main_thread;
+#endif
+
 #ifdef CONFIG_BOOTARGS
 	extern int main(int argc, char **argv);
 	extern char **sys_boot_prepare_main_args(int *argc);
