diff --git a/lib/os/clock.c b/lib/os/clock.c
index 1a902dd4445..e4d98de11a9 100644
--- a/lib/os/clock.c
+++ b/lib/os/clock.c
@@ -210,9 +210,7 @@ int z_vrfy_sys_clock_nanosleep(int clock_id, int flags, const struct timespec *r
 #include <zephyr/ztest.h>
 static void reset_clock_offset(void)
 {
-	K_SPINLOCK(&rt_clock_offset_lock) {
-		rt_clock_offset = (struct timespec){0};
-	}
+	rt_clock_offset = (struct timespec){0};
 }
 
 static void clock_offset_reset_rule_after(const struct ztest_unit_test *test, void *data)
