diff --git a/lib/libc/common/include/machine/_threads.h b/lib/libc/common/include/machine/_threads.h
index cebce717c73..1ce346953ef 100644
--- a/lib/libc/common/include/machine/_threads.h
+++ b/lib/libc/common/include/machine/_threads.h
@@ -18,7 +18,7 @@ typedef int mtx_t;
 typedef int thrd_t;
 typedef int tss_t;
 typedef struct {
-	char flag;
+	unsigned long flag;
 } once_flag;
 
 #ifdef __cplusplus
