diff --git a/lib/libc/newlib/include/netinet/icmp6.h b/lib/libc/newlib/include/netinet/icmp6.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/newlib/include/netinet/icmp6.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_ICMP6_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_ICMP6_H_
+
+#include <zephyr/posix/netinet/icmp6.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_ICMP6_H_ */
diff --git a/lib/libc/newlib/include/netinet/ip.h b/lib/libc/newlib/include/netinet/ip.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/newlib/include/netinet/ip.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_IP_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_IP_H_
+
+#include <zephyr/posix/netinet/ip.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_IP_H_ */
diff --git a/lib/libc/newlib/include/netinet/ip_icmp.h b/lib/libc/newlib/include/netinet/ip_icmp.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/newlib/include/netinet/ip_icmp.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_IP_ICMP_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_IP_ICMP_H_
+
+#include <zephyr/posix/netinet/ip_icmp.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETINET_IP_ICMP_H_ */
diff --git a/lib/libc/picolibc/include/netinet/icmp6.h b/lib/libc/picolibc/include/netinet/icmp6.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/picolibc/include/netinet/icmp6.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_ICMP6_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_ICMP6_H_
+
+#include <zephyr/posix/netinet/icmp6.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_ICMP6_H_ */
diff --git a/lib/libc/picolibc/include/netinet/ip.h b/lib/libc/picolibc/include/netinet/ip.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/picolibc/include/netinet/ip.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_IP_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_IP_H_
+
+#include <zephyr/posix/netinet/ip.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_IP_H_ */
diff --git a/lib/libc/picolibc/include/netinet/ip_icmp.h b/lib/libc/picolibc/include/netinet/ip_icmp.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/picolibc/include/netinet/ip_icmp.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_IP_ICMP_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_IP_ICMP_H_
+
+#include <zephyr/posix/netinet/ip_icmp.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETINET_IP_ICMP_H_ */
