diff --git a/include/zephyr/kernel.h b/include/zephyr/kernel.h
index 9191370a93c..e376486c479 100644
--- a/include/zephyr/kernel.h
+++ b/include/zephyr/kernel.h
@@ -3349,6 +3349,15 @@ struct k_condvar {
 #ifdef CONFIG_OBJ_CORE_CONDVAR
 	struct k_obj_core  obj_core;
 #endif
+
+#ifdef CONFIG_SYS_THREAD
+	/*
+	 * bits 0-2: clock id
+	 * bit 3: initialized
+	 * bit 4: pshared
+	 */
+	uint32_t flags;
+#endif
 };
 
 #define Z_CONDVAR_INITIALIZER(obj)                                             \
