diff --git a/lib/hash/Kconfig.hash_map b/lib/hash/Kconfig.hash_map
index d36a4e12453..b02d2db093e 100644
--- a/lib/hash/Kconfig.hash_map
+++ b/lib/hash/Kconfig.hash_map
@@ -37,15 +37,15 @@ config SYS_HASH_MAP_OA_LP
 	  contiguous allocation which improves performance on systems with
 	  memory caching.
 
-config SYS_HASH_MAP_CXX
-	bool "C++ Hashmap"
-	select CPP
-	select REQUIRES_FULL_LIBCPP
-	select CPP_EXCEPTIONS
-	help
-	  This enables a C wrapper around the C++ std::unordered_map.
-
-	  It is mainly used for benchmarking purposes.
+# config SYS_HASH_MAP_CXX
+# 	bool "C++ Hashmap"
+# 	select CPP
+# 	select REQUIRES_FULL_LIBCPP
+# 	select CPP_EXCEPTIONS
+# 	help
+# 	  This enables a C wrapper around the C++ std::unordered_map.
+#
+# 	  It is mainly used for benchmarking purposes.
 
 choice SYS_HASH_MAP_CHOICE
 	prompt "Default hashmap implementation"
diff --git a/lib/os/zvfs/Kconfig b/lib/os/zvfs/Kconfig
index 794c15142db..117944dbdc6 100644
--- a/lib/os/zvfs/Kconfig
+++ b/lib/os/zvfs/Kconfig
@@ -32,6 +32,10 @@ config ZVFS_EVENTFD_MAX
 	help
 	  The maximum number of supported event file descriptors.
 
+config ZVFS_OPEN_ADD_SIZE_EVENTFD
+       int
+       default ZVFS_EVENTFD_MAX
+
 endif # ZVFS_EVENTFD
 
 config ZVFS_POLL
