diff --git a/subsys/net/lib/sockets/sockets.c b/subsys/net/lib/sockets/sockets.c
--- a/subsys/net/lib/sockets/sockets.c
+++ b/subsys/net/lib/sockets/sockets.c
@@ -1199,3 +1199,24 @@
 
 	return 0;
 }
+
+#if !defined(CONFIG_POSIX_NETWORKING)
+/*
+ * this is defined here, weakly, since mbedtls assumes it is available based on the presence of
+ * the <arpa/inet.h> header.
+ *
+ * ```
+ * $HOME/zephyrproject/modules/crypto/mbedtls/library/x509_crt.c:2872: undefined reference to
+ * `inet_pton'
+ *
+ * #if __has_include(<arpa/inet.h>)
+ * #include <arpa/inet.h>
+ * #endif
+ * ```
+ */
+
+__weak int inet_pton(int af, const char *ZRESTRICT src, void *ZRESTRICT dst)
+{
+	return zsock_inet_pton(af, src, dst);
+}
+#endif
