diff --git a/include/zephyr/net/net_ip.h b/include/zephyr/net/net_ip.h
--- a/include/zephyr/net/net_ip.h
+++ b/include/zephyr/net/net_ip.h
@@ -1252,7 +1252,7 @@
 /** @cond INTERNAL_HIDDEN */
 static inline bool net_ipv6_is_ll_addr_raw(const uint8_t *addr)
 {
-	return UNALIGNED_GET((uint16_t *)addr) == net_htons(0xFE80);
+	return (UNALIGNED_GET((uint16_t *)addr) & net_htons(0xFFC0)) == net_htons(0xFE80);
 }
 /** @endcond */
 
@@ -1277,7 +1277,7 @@
  */
 static inline bool net_ipv6_is_sl_addr(const struct net_in6_addr *addr)
 {
-	return UNALIGNED_GET(&addr->s6_addr16[0]) == net_htons(0xFEC0);
+	return (UNALIGNED_GET(&addr->s6_addr16[0]) & net_htons(0xFFC0)) == net_htons(0xFEC0);
 }
 
 
