diff --git a/include/zephyr/net/net_ip.h b/include/zephyr/net/net_ip.h
index 84980130cdc..62021637d29 100644
--- a/include/zephyr/net/net_ip.h
+++ b/include/zephyr/net/net_ip.h
@@ -465,10 +465,16 @@ extern const struct zsock_in6_addr zsock_in6addr_loopback;
 #define ZSOCK_INADDR_BROADCAST 0xffffffff
 
 /** IPv4 any-address initializer */
-#define INADDR_ANY_INIT { { { ZSOCK_INADDR_ANY } } }
+#define ZSOCK_INADDR_ANY_INIT { { { ZSOCK_INADDR_ANY } } }
+#ifndef INADDR_ANY_INIT
+#define INADDR_ANY_INIT ZSOCK_INADDR_ANY_INIT
+#endif
 
 /** IPv4 loopback address initializer */
-#define INADDR_LOOPBACK_INIT  { { { 127, 0, 0, 1 } } }
+#define ZSOCK_INADDR_LOOPBACK_INIT  { { { 127, 0, 0, 1 } } }
+#ifndef INADDR_LOOPBACK_INIT
+#define INADDR_LOOPBACK_INIT ZSOCK_INADDR_LOOPBACK_INIT
+#endif
 
 /** Max length of the IPv4 address as a string. Defined by POSIX. */
 #define ZSOCK_INET_ADDRSTRLEN 16
