diff --git a/subsys/net/ip/utils.c b/subsys/net/ip/utils.c
index d23deb97cc5..4d9983681f5 100644
--- a/subsys/net/ip/utils.c
+++ b/subsys/net/ip/utils.c
@@ -329,7 +329,7 @@ char *z_vrfy_net_addr_ntop(net_sa_family_t family, const void *src,
 		return 0;
 	}

-	out = z_impl_net_addr_ntop(family, addr, str, sizeof(str));
+	out = z_impl_net_addr_ntop(family, addr, str, MIN(size, sizeof(str)));
 	if (!out) {
 		return 0;
 	}
