diff --git a/include/zephyr/net/net_ip.h b/include/zephyr/net/net_ip.h
index e5673a14578..e0f11f00279 100644
--- a/include/zephyr/net/net_ip.h
+++ b/include/zephyr/net/net_ip.h
@@ -396,9 +396,21 @@ struct zsock_cmsghdr {
 };
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef _IOVEC_DECLARED
 #define iovec           zsock_iovec
+#define _IOVEC_DECLARED
+#define __iovec_defined
+#endif
+#ifndef _MSGHDR_DECLARED
 #define msghdr          zsock_msghdr
+#define _MSGHDR_DECLARED
+#define __msghdr_defined
+#endif
+#ifndef _CMSGHDR_DECLARED
 #define cmsghdr         zsock_cmsghdr
+#define _CMSGHDR_DECLARED
+#define __cmsghdr_defined
+#endif
 
 /** @cond INTERNAL_HIDDEN */
 
@@ -571,30 +583,47 @@ struct zsock_sockaddr_un {
 };
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef _SOCKADDR_DECLARED
 #define sockaddr          zsock_sockaddr
+#define _SOCKADDR_DECLARED
+#define __sockaddr_defined
+#endif
 #ifndef _SOCKADDR_IN_DECLARED
 #define sockaddr_in       zsock_sockaddr_in
+#define _SOCKADDR_IN_DECLARED
+#define __sockaddr_in_defined
 #endif
 #ifndef _SOCKADDR_IN6_DECLARED
 #define sockaddr_in6      zsock_sockaddr_in6
+#define _SOCKADDR_IN6_DECLARED
+#define __sockaddr_in6_defined
 #endif
 #ifndef _SOCKADDR_STORAGE_DECLARED
 #define sockaddr_storage  zsock_sockaddr_storage
+#define _SOCKADDR_STORAGE_DECLARED
+#define __sockaddr_storage_defined
 #endif
 #ifndef _SOCKADDR_UN_DECLARED
 #define sockaddr_un       zsock_sockaddr_un
+#define _SOCKADDR_UN_DECLARED
 #endif
 #ifndef _SOCKADDR_LL_DECLARED
 #define sockaddr_ll       zsock_sockaddr_ll
+#define _SOCKADDR_LL_DECLARED
 #endif
 #ifndef _SOCKADDR_CAN_DECLARED
 #define sockaddr_can      zsock_sockaddr_can
+#define _SOCKADDR_CAN_DECLARED
 #endif
 #ifndef _IN_ADDR_DECLARED
 #define in_addr           zsock_in_addr
+#define _IN_ADDR_DECLARED
+#define __in_addr_defined
 #endif
 #ifndef _IN6_ADDR_DECLARED
 #define in6_addr          zsock_in6_addr
+#define _IN6_ADDR_DECLARED
+#define __in6_addr_defined
 #endif
 
 struct net_addr {
@@ -611,6 +640,7 @@ struct net_addr {
 extern const struct zsock_in6_addr in6addr_any;
 /** A pointer to IPv6 loopback address (::1) */
 extern const struct zsock_in6_addr in6addr_loopback;
+#define _POSIX_NET_IN6ADDR_DECLARED
 #endif
 #ifndef zsock_in6addr_any
 #define zsock_in6addr_any in6addr_any
diff --git a/include/zephyr/net/socket.h b/include/zephyr/net/socket.h
index 4b1437342c9..9e9f63ead1a 100644
--- a/include/zephyr/net/socket.h
+++ b/include/zephyr/net/socket.h
@@ -358,6 +358,8 @@ struct zsock_addrinfo {
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
 #ifndef _ADDRINFO_DECLARED
 #define addrinfo zsock_addrinfo
+#define _ADDRINFO_DECLARED
+#define __addrinfo_defined
 #endif
 
 /**
@@ -1293,6 +1295,8 @@ struct zsock_ipv6_mreq {
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
 #ifndef _IPV6_MREQ_DECLARED
 #define ipv6_mreq zsock_ipv6_mreq
+#define _IPV6_MREQ_DECLARED
+#define __ipv6_mreq_defined
 #endif
 
 /** For getsockopt(), retrieve the current known IPv6 path MTU of the given socket.
