diff --git a/include/zephyr/net/net_ip.h b/include/zephyr/net/net_ip.h
index e0f11f00279..62d47b07ff5 100644
--- a/include/zephyr/net/net_ip.h
+++ b/include/zephyr/net/net_ip.h
@@ -142,7 +142,9 @@ enum net_sock_type {
 };
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef IPPROTO_IP
 #define IPPROTO_IP      ZSOCK_IPPROTO_IP
+#endif
 #ifndef IPPROTO_ICMP
 #define IPPROTO_ICMP    ZSOCK_IPPROTO_ICMP
 #endif
diff --git a/include/zephyr/net/socket.h b/include/zephyr/net/socket.h
index 9e9f63ead1a..8963d8f87b3 100644
--- a/include/zephyr/net/socket.h
+++ b/include/zephyr/net/socket.h
@@ -829,7 +829,9 @@ __syscall int z_zsock_getaddrinfo_internal(const char *host,
 #define ZSOCK_AI_NUMERICSERV 0x400
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef AI_PASSIVE
 #define AI_PASSIVE     ZSOCK_AI_PASSIVE
+#endif
 #ifndef AI_CANONNAME
 #define AI_CANONNAME   ZSOCK_AI_CANONNAME
 #endif
@@ -902,7 +904,9 @@ const char *zsock_gai_strerror(int errcode);
 #define ZSOCK_NI_DGRAM 16
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef NI_NUMERICHOST
 #define NI_NUMERICHOST ZSOCK_NI_NUMERICHOST
+#endif
 #ifndef NI_NUMERICSERV
 #define NI_NUMERICSERV ZSOCK_NI_NUMERICSERV
 #endif
@@ -1036,7 +1040,9 @@ struct ifreq {
 #define ZSOCK_SO_TXTIME 61
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef SOL_SOCKET
 #define SOL_SOCKET     ZSOCK_SOL_SOCKET
+#endif
 #ifndef SO_DEBUG
 #define SO_DEBUG       ZSOCK_SO_DEBUG
 #endif
@@ -1146,7 +1152,9 @@ struct ifreq {
 #define ZSOCK_TCP_KEEPCNT 4
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef TCP_NODELAY
 #define TCP_NODELAY    ZSOCK_TCP_NODELAY
+#endif
 #ifndef TCP_KEEPIDLE
 #define TCP_KEEPIDLE   ZSOCK_TCP_KEEPIDLE
 #endif
@@ -1435,7 +1443,9 @@ struct in6_pktinfo {
 #define ZSOCK_IN6_IS_ADDR_MC_ORGLOCAL(addr) net_ipv6_is_addr_mcast_org(addr)
 
 /* TEMP_POSIX_ZSOCK_NS_WORKAROUND */
+#ifndef IN6_IS_ADDR_UNSPECIFIED
 #define IN6_IS_ADDR_UNSPECIFIED(addr)  ZSOCK_IN6_IS_ADDR_UNSPECIFIED(addr)
+#endif
 #ifndef IN6_IS_ADDR_LOOPBACK
 #define IN6_IS_ADDR_LOOPBACK(addr)     ZSOCK_IN6_IS_ADDR_LOOPBACK(addr)
 #endif
