diff --git a/subsys/net/lib/sockets/getaddrinfo.c b/subsys/net/lib/sockets/getaddrinfo.c
--- a/subsys/net/lib/sockets/getaddrinfo.c
+++ b/subsys/net/lib/sockets/getaddrinfo.c
@@ -356,7 +356,7 @@
 #endif /* defined(CONFIG_DNS_RESOLVER) */
 
 #if defined(CONFIG_NET_IP)
-static int try_resolve_literal_addr(const char *host, const char *service,
+int try_resolve_literal_addr(const char *host, const char *service,
 				    const struct zsock_addrinfo *hints,
 				    struct zsock_addrinfo *res)
 {
