diff --git a/subsys/net/lib/ptp/transport.c b/subsys/net/lib/ptp/transport.c
index 1c5fc56a583..67483e2c3e7 100644
--- a/subsys/net/lib/ptp/transport.c
+++ b/subsys/net/lib/ptp/transport.c
@@ -336,7 +336,7 @@ static int transport_l2_open(struct net_if *iface)
 static int transport_send_udp(int socket, int port, void *buf, int length,
 			      struct net_sockaddr *addr)
 {
-	struct net_sockaddr_storage m_addr_storage;
+	struct net_sockaddr_storage m_addr_storage = { 0 };
 	struct net_sockaddr *m_addr = net_sad(&m_addr_storage);
 	net_socklen_t addrlen;
 	int cnt;
diff --git a/subsys/net/lib/sntp/sntp_server.c b/subsys/net/lib/sntp/sntp_server.c
index 0cec34dc6e7..2cb87f0b693 100644
--- a/subsys/net/lib/sntp/sntp_server.c
+++ b/subsys/net/lib/sntp/sntp_server.c
@@ -50,7 +50,7 @@ static void sntp_service_handler(struct net_socket_service_event *pev)
 {
 	struct zsock_pollfd *pfd = &pev->event;
 	int sock = pfd->fd;
-	struct net_sockaddr_storage addr;
+	struct net_sockaddr_storage addr = { 0 };
 	net_socklen_t addrlen = sizeof(addr);
 	struct sntp_pkt request, reply;
 	struct timespec now;
