diff --git a/include/zephyr/net/midi2.h b/include/zephyr/net/midi2.h
index 4af8edc3258..547db8f07b7 100644
--- a/include/zephyr/net/midi2.h
+++ b/include/zephyr/net/midi2.h
@@ -31,7 +31,7 @@
 #include <stdint.h>
 #include <zephyr/audio/midi.h>
 #include <zephyr/net/socket.h>
-#include <zephyr/posix/poll.h>
+#include <zephyr/sys/fdtable.h>
 
 /**
  * Size, in bytes, of the nonce sent to the client for authentication
@@ -214,7 +214,7 @@ struct netmidi2_ep {
 		struct sockaddr_in6 addr6;
 	};
 	/** The listening socket wrapped in a poll descriptor */
-	struct pollfd pollsock;
+	struct zvvfs_pollfd pollsock;
 	/** The function to call when data is received from a client */
 	void (*rx_packet_cb)(struct netmidi2_session *session,
 			     const struct midi_ump ump);
