diff --git a/lib/libc/newlib/include/sys/un.h b/lib/libc/newlib/include/sys/un.h
new file mode 100644
index 00000000000..43d1ef06ac3
--- /dev/null
+++ b/lib/libc/newlib/include/sys/un.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_SYS_UN_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_SYS_UN_H_
+
+#include <zephyr/posix/sys/un.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_SYS_UN_H_ */
diff --git a/lib/libc/picolibc/include/sys/un.h b/lib/libc/picolibc/include/sys/un.h
new file mode 100644
index 00000000000..2b69b116765
--- /dev/null
+++ b/lib/libc/picolibc/include/sys/un.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_SYS_UN_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_SYS_UN_H_
+
+#include <zephyr/posix/sys/un.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_SYS_UN_H_ */
