diff --git a/lib/libc/picolibc/include/netpacket/packet.h b/lib/libc/picolibc/include/netpacket/packet.h
new file mode 100644
index 00000000000..c8f3e2a1b4d
--- /dev/null
+++ b/lib/libc/picolibc/include/netpacket/packet.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETPACKET_PACKET_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETPACKET_PACKET_H_
+
+#include <zephyr/posix/netpacket/packet.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETPACKET_PACKET_H_ */
