diff --git a/include/zephyr/posix/netpacket/packet.h b/include/zephyr/posix/netpacket/packet.h
new file mode 100644
--- /dev/null
+++ b/include/zephyr/posix/netpacket/packet.h
@@ -0,0 +1,17 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Packet socket address (<netpacket/packet.h>)
+ */
+
+#ifndef ZEPHYR_INCLUDE_POSIX_NETPACKET_PACKET_H_
+#define ZEPHYR_INCLUDE_POSIX_NETPACKET_PACKET_H_
+
+#include <zephyr/net/net_ip.h>
+
+#endif /* ZEPHYR_INCLUDE_POSIX_NETPACKET_PACKET_H_ */
diff --git a/lib/libc/newlib/include/netpacket/packet.h b/lib/libc/newlib/include/netpacket/packet.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/newlib/include/netpacket/packet.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETPACKET_PACKET_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETPACKET_PACKET_H_
+
+#include <zephyr/posix/netpacket/packet.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_NETPACKET_PACKET_H_ */
diff --git a/lib/libc/picolibc/include/netpacket/packet.h b/lib/libc/picolibc/include/netpacket/packet.h
new file mode 100644
--- /dev/null
+++ b/lib/libc/picolibc/include/netpacket/packet.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETPACKET_PACKET_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETPACKET_PACKET_H_
+
+#include <zephyr/posix/netpacket/packet.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_NETPACKET_PACKET_H_ */
