diff --git a/lib/libc/arcmwdt/include/pthread.h b/lib/libc/arcmwdt/include/pthread.h
index ec93137904f..cdb75b72586 100644
--- a/lib/libc/arcmwdt/include/pthread.h
+++ b/lib/libc/arcmwdt/include/pthread.h
@@ -22,7 +22,7 @@
 	}
 
 /* The minimum allowable stack size */
-#define PTHREAD_STACK_MIN K_KERNEL_STACK_LEN(0)
+#define PTHREAD_STACK_MIN CONFIG_IDLE_STACK_SIZE
 
 void __z_pthread_cleanup_push(void *cleanup[3], void (*routine)(void *arg), void *arg);
 void __z_pthread_cleanup_pop(int execute);
diff --git a/lib/libc/newlib/include/pthread.h b/lib/libc/newlib/include/pthread.h
index 448836f0ccf..f5875bf8af1 100644
--- a/lib/libc/newlib/include/pthread.h
+++ b/lib/libc/newlib/include/pthread.h
@@ -22,7 +22,7 @@
 	}
 
 /* The minimum allowable stack size */
-#define PTHREAD_STACK_MIN K_KERNEL_STACK_LEN(0)
+#define PTHREAD_STACK_MIN CONFIG_IDLE_STACK_SIZE
 
 void __z_pthread_cleanup_push(void *cleanup[3], void (*routine)(void *arg), void *arg);
 void __z_pthread_cleanup_pop(int execute);
diff --git a/lib/libc/picolibc/include/pthread.h b/lib/libc/picolibc/include/pthread.h
index fc5c6fd545a..8c368699222 100644
--- a/lib/libc/picolibc/include/pthread.h
+++ b/lib/libc/picolibc/include/pthread.h
@@ -30,7 +30,7 @@
 	}
 
 /* The minimum allowable stack size */
-#define PTHREAD_STACK_MIN K_KERNEL_STACK_LEN(0)
+#define PTHREAD_STACK_MIN CONFIG_IDLE_STACK_SIZE
 
 void __z_pthread_cleanup_push(void *cleanup[3], void (*routine)(void *arg), void *arg);
 void __z_pthread_cleanup_pop(int execute);
