diff --git a/arch/riscv/core/pmp.c b/arch/riscv/core/pmp.c
index 49828513d71..5ce5f42227c 100644
--- a/arch/riscv/core/pmp.c
+++ b/arch/riscv/core/pmp.c
@@ -500,9 +500,9 @@ static inline unsigned int z_riscv_pmp_thread_init(unsigned long *pmp_addr,
 	 * Retrieve the pmpaddr value matching the last global PMP slot.
 	 * This is so that set_pmp_entry() can safely attempt TOR with it.
 	 */
-	pmp_addr[global_pmp_end_index - 1] = global_pmp_last_addr;
+	pmp_addr[MAX(1, global_pmp_end_index) - 1] = global_pmp_last_addr;
 
-	return global_pmp_end_index;
+	return MAX(1, global_pmp_end_index);
 }
 #endif
 
