diff --git a/lib/libc/newlib/include/signal.h b/lib/libc/newlib/include/signal.h
index 16dba56b74d..d50404b4d70 100644
--- a/lib/libc/newlib/include/signal.h
+++ b/lib/libc/newlib/include/signal.h
@@ -16,7 +16,7 @@
  * reserve 32 (cancel) and 33, with application RT signals from 34.
  */
 #undef SIGRTMIN
-#define SIGRTMIN 34
+#define SIGRTMIN (CONFIG_THREAD_CANCEL_SIGNAL_NUMBER + 2)
 #if defined(_POSIX_REALTIME_SIGNALS) || defined(__DOXYGEN__)
 #undef SIGRTMAX
 #define SIGRTMAX (SIGRTMIN + CONFIG_POSIX_RTSIG_MAX)
diff --git a/lib/libc/picolibc/include/signal.h b/lib/libc/picolibc/include/signal.h
index e4914c9a95a..2d73f69d5ec 100644
--- a/lib/libc/picolibc/include/signal.h
+++ b/lib/libc/picolibc/include/signal.h
@@ -18,7 +18,7 @@
  * reserve 32 (cancel) and 33, with application RT signals from 34.
  */
 #undef SIGRTMIN
-#define SIGRTMIN 34
+#define SIGRTMIN (CONFIG_THREAD_CANCEL_SIGNAL_NUMBER + 2)
 #if defined(_POSIX_REALTIME_SIGNALS) || defined(__DOXYGEN__)
 #undef SIGRTMAX
 #define SIGRTMAX (SIGRTMIN + CONFIG_POSIX_RTSIG_MAX)
