diff --git a/include/zephyr/net/socket.h b/include/zephyr/net/socket.h
--- a/include/zephyr/net/socket.h
+++ b/include/zephyr/net/socket.h
@@ -1282,36 +1282,4 @@
 /**
  * @}
  */
-
-/* Avoid circular loops with POSIX socket headers.
- * We have these includes here so that we do not need
- * to change the applications that were only including
- * zephyr/net/socket.h header file.
- *
- * Additionally, if non-zephyr-prefixed headers are used here,
- * native_sim pulls in those from the host rather than Zephyr's.
- */
-#if defined(CONFIG_POSIX_NETWORKING)
-
-/*
- * Note: this is required to avoid globally defining _POSIX_C_SOURCE and _XOPEN_SOURCE. The
- * dependency loop should be fixed with #97050 or adjacent PRs.
- */
-#undef _POSIX_C_SOURCE
-#define _POSIX_C_SOURCE 200809L
-#undef _XOPEN_SOURCE
-#define _XOPEN_SOURCE 700
-
-#include <sys/stat.h>
-#include <arpa/inet.h>
-#include <netdb.h>
-#include <unistd.h>
-#include <poll.h>
-#include <sys/socket.h>
-
-#undef _POSIX_C_SOURCE
-#undef _XOPEN_SOURCE
-
-#endif /* CONFIG_POSIX_NETWORKING */
-
 #endif /* ZEPHYR_INCLUDE_NET_SOCKET_H_ */
