diff --git a/lib/os/thread.c b/lib/os/thread.c
index e421604a31c..b37ee0b5eb6 100644
--- a/lib/os/thread.c
+++ b/lib/os/thread.c
@@ -662,7 +662,7 @@ static int sys_thread_create_common(struct k_thread **th, k_thread_stack_t *stac
 		k_object_access_grant(*th, _current);
 		k_object_uninit(*th);
 		z_vrfy_k_thread_create(*th, stack, stack_size, entry, p1, p2, p3, prio, options,
-				       K_NO_WAIT);
+				       delay);
 		return 0;
 	}
 #endif
