diff --git a/arch/x86/core/Kconfig.ia32 b/arch/x86/core/Kconfig.ia32
index f24e7d37758..7a39571d9e0 100644
--- a/arch/x86/core/Kconfig.ia32
+++ b/arch/x86/core/Kconfig.ia32
@@ -67,7 +67,7 @@ config X86_STACK_PROTECTION
 	select SET_GDT
 	select GDT_DYNAMIC
 	select X86_ENABLE_TSS
-	imply THREAD_STACK_MEM_MAPPED if !DEMAND_PAGING
+	imply THREAD_STACK_MEM_MAPPED if !(DEMAND_PAGING || SYS_THREAD)
 	help
 	  This option leverages the MMU to cause a system fatal error if the
 	  bounds of the current process stack are overflowed. This is done
diff --git a/arch/x86/core/Kconfig.intel64 b/arch/x86/core/Kconfig.intel64
index 0f73341db31..186627726fa 100644
--- a/arch/x86/core/Kconfig.intel64
+++ b/arch/x86/core/Kconfig.intel64
@@ -75,7 +75,7 @@ config X86_STACK_PROTECTION
 	bool
 	default y if HW_STACK_PROTECTION
 	select THREAD_STACK_INFO
-	imply THREAD_STACK_MEM_MAPPED
+	imply THREAD_STACK_MEM_MAPPED if ! SYS_THREAD
 	help
 	  This option leverages the MMU to cause a system fatal error if the
 	  bounds of the current process stack are overflowed. This is done
diff --git a/include/zephyr/kernel/thread.h b/include/zephyr/kernel/thread.h
index a3b822cd140..f186e9967db 100644
--- a/include/zephyr/kernel/thread.h
+++ b/include/zephyr/kernel/thread.h
@@ -198,7 +198,7 @@ struct _thread_stack_info {
 	 */
 	size_t delta;

-#if defined(CONFIG_THREAD_STACK_MEM_MAPPED)
+#if defined(CONFIG_THREAD_STACK_MEM_MAPPED) || defined(CONFIG_SYS_THREAD_STACK_ALLOC_ANON)
 	struct {
 		/** Base address of the memory mapped thread stack */
 		k_thread_stack_t *addr;
diff --git a/kernel/thread.c b/kernel/thread.c
index 6a8e600301d..9517dcd2bfd 100644
--- a/kernel/thread.c
+++ b/kernel/thread.c
@@ -1293,18 +1293,20 @@ void defer_thread_cleanup(struct k_thread *thread)
 	 */
 	thread_to_cleanup = thread;

-#ifdef CONFIG_THREAD_STACK_MEM_MAPPED
+#if defined(CONFIG_THREAD_STACK_MEM_MAPPED) || defined(CONFIG_SYS_THREAD_STACK_ALLOC_ANON)
 	/* Note that the permission of the stack should have been
 	 * stripped of user thread access due to the thread having
 	 * already exited from a memory domain. That is done via
 	 * k_thread_abort().
 	 */

+#ifdef CONFIG_THREAD_ABORT_NEED_CLEANUP
 	/* Stash the address and size so the region can be unmapped
 	 * later.
 	 */
 	thread_cleanup_stack_addr = thread->stack_info.mapped.addr;
 	thread_cleanup_stack_sz = thread->stack_info.mapped.sz;
+#endif

 	/* The stack is now considered un-usable. This should prevent any functions
 	 * from looking directly into the mapped stack if they are made to be aware
@@ -1330,7 +1332,7 @@ void do_thread_cleanup(struct k_thread *thread)

 		thread_cleanup_stack_addr = NULL;
 	}
-#endif /* CONFIG_THREAD_STACK_MEM_MAPPED */
+#endif /* CONFIG_THREAD_STACK_MEM_MAPPED || CONFIG_SYS_THREAD_STACK_ALLOC_ANON */
 }

 void k_thread_abort_cleanup(struct k_thread *thread)
