diff --git a/boards/qemu/x86/qemu_x86_tiny.ld b/boards/qemu/x86/qemu_x86_tiny.ld
index b94d0fb319d..a05cd1909b1 100644
--- a/boards/qemu/x86/qemu_x86_tiny.ld
+++ b/boards/qemu/x86/qemu_x86_tiny.ld
@@ -804,8 +804,6 @@ SECTIONS
 
 	_image_ram_all = (KERNEL_BASE_ADDR + KERNEL_RAM_SIZE) - _image_ram_start;
 
-	z_mapped_size = z_mapped_end - z_mapped_start;
-
 #ifndef LINKER_ZEPHYR_FINAL
 	/* static interrupts */
 	SECTION_PROLOGUE(intList,,)
@@ -836,6 +834,9 @@ SECTIONS
 
 #include <zephyr/linker/ram-end.ld>
 
+	. = ALIGN(CONFIG_MMU_PAGE_SIZE);
+	z_mapped_size = . - z_mapped_start;
+
     GROUP_END(RAMABLE_REGION)
 
 #include <zephyr/linker/debug-sections.ld>
diff --git a/include/zephyr/arch/x86/ia32/linker.ld b/include/zephyr/arch/x86/ia32/linker.ld
index 6f238902a83..b6350f27de5 100644
--- a/include/zephyr/arch/x86/ia32/linker.ld
+++ b/include/zephyr/arch/x86/ia32/linker.ld
@@ -550,7 +550,6 @@ SECTION_PROLOGUE(.x86shadowstack.arr,,)
 	_image_ram_all = (KERNEL_BASE_ADDR + KERNEL_RAM_SIZE) - _image_ram_start;
 
 	z_mapped_end = .;
-	z_mapped_size = z_mapped_end - z_mapped_start;
 
 #ifndef LINKER_ZEPHYR_FINAL
 	/* static interrupts */
@@ -580,6 +579,9 @@ SECTION_PROLOGUE(.x86shadowstack.arr,,)
 
 #include <zephyr/linker/ram-end.ld>
 
+	. = ALIGN(CONFIG_MMU_PAGE_SIZE);
+	z_mapped_size = . - z_mapped_start;
+
     GROUP_END(RAMABLE_REGION)
 
 #include <zephyr/linker/debug-sections.ld>
