diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index e3b748cc714..03a8c6e4419 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -332,6 +332,13 @@ config X86_MMU
 	  and creates a set of page tables at boot time that is runtime-
 	  mutable.
 
+if X86_MMU
+
+config MMU_PAGE_SIZE
+	default 0x1000
+
+endif
+
 config X86_COMMON_PAGE_TABLE
 	bool "Use a single page table for all threads"
 	default n
