diff --git a/boards/qemu/x86/qemu_x86_tiny.ld b/boards/qemu/x86/qemu_x86_tiny.ld
index 43b474b955b..c2dee17260a 100644
--- a/boards/qemu/x86/qemu_x86_tiny.ld
+++ b/boards/qemu/x86/qemu_x86_tiny.ld
@@ -801,8 +801,6 @@ SECTIONS

 	_image_ram_all = (KERNEL_BASE_ADDR + KERNEL_RAM_SIZE) - _image_ram_start;

-	z_mapped_size = z_mapped_end - z_mapped_start;
-
 #ifndef LINKER_ZEPHYR_FINAL
 	/* static interrupts */
 	SECTION_PROLOGUE(intList,,)
@@ -833,6 +831,9 @@ SECTIONS

 #include <zephyr/linker/ram-end.ld>

+	. = ALIGN(CONFIG_MMU_PAGE_SIZE);
+	z_mapped_size = . - z_mapped_start;
+
     GROUP_END(RAMABLE_REGION)

 #include <zephyr/linker/debug-sections.ld>
diff --git a/include/zephyr/arch/x86/ia32/linker.ld b/include/zephyr/arch/x86/ia32/linker.ld
index 6423ab35f67..b0222192ef6 100644
--- a/include/zephyr/arch/x86/ia32/linker.ld
+++ b/include/zephyr/arch/x86/ia32/linker.ld
@@ -545,7 +545,6 @@ SECTION_PROLOGUE(.x86shadowstack.arr,,)
 	_image_ram_all = (KERNEL_BASE_ADDR + KERNEL_RAM_SIZE) - _image_ram_start;

 	z_mapped_end = .;
-	z_mapped_size = z_mapped_end - z_mapped_start;

 #ifndef LINKER_ZEPHYR_FINAL
 	/* static interrupts */
@@ -575,6 +574,9 @@ SECTION_PROLOGUE(.x86shadowstack.arr,,)

 #include <zephyr/linker/ram-end.ld>

+	. = ALIGN(CONFIG_MMU_PAGE_SIZE);
+	z_mapped_size = . - z_mapped_start;
+
     GROUP_END(RAMABLE_REGION)

 #include <zephyr/linker/debug-sections.ld>
