diff --git a/lib/libc/minimal/include/getopt.h b/lib/libc/minimal/include/getopt.h
new file mode 100644
index 00000000000..c28b1d9d9fa
--- /dev/null
+++ b/lib/libc/minimal/include/getopt.h
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_GETOPT_H_
+#define ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_GETOPT_H_
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+#define no_argument       0
+#define required_argument 1
+#define optional_argument 2
+
+struct option {
+	const char *name;
+	int has_arg;
+	int *flag;
+	int val;
+};
+
+extern char *optarg;
+extern int opterr, optind, optopt;
+
+int getopt(int argc, char *const argv[], const char *optstring);
+int getopt_long(int argc, char *const argv[], const char *optstring,
+	const struct option *longopts, int *longindex);
+int getopt_long_only(int argc, char *const argv[], const char *optstring,
+	const struct option *longopts, int *longindex);
+
+#if defined(__cplusplus)
+}
+#endif
+
+#endif /* ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_GETOPT_H_ */
diff --git a/lib/libc/newlib/include/getopt.h b/lib/libc/newlib/include/getopt.h
new file mode 100644
index 00000000000..49b0e9acbe0
--- /dev/null
+++ b/lib/libc/newlib/include/getopt.h
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_GETOPT_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_GETOPT_H_
+
+#include_next <getopt.h>
+
+#include <zephyr/getopt.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_GETOPT_H_ */
diff --git a/lib/libc/picolibc/include/getopt.h b/lib/libc/picolibc/include/getopt.h
new file mode 100644
index 00000000000..326c864673e
--- /dev/null
+++ b/lib/libc/picolibc/include/getopt.h
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_GETOPT_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_GETOPT_H_
+
+#include_next <getopt.h>
+
+#include <zephyr/getopt.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_GETOPT_H_ */
