diff --git a/lib/libc/arcmwdt/include/fcntl.h b/lib/libc/arcmwdt/include/fcntl.h
index 84c91e118d7..8073e2182b1 100644
--- a/lib/libc/arcmwdt/include/fcntl.h
+++ b/lib/libc/arcmwdt/include/fcntl.h
@@ -8,21 +8,36 @@
 #define ZEPHYR_LIB_LIBC_ARCMWDT_INCLUDE_FCNTL_H_

 #include_next <fcntl.h>
+#include <zephyr/sys/fdtable.h>

 #define F_DUPFD 0
-#define F_GETFL 3
-#define F_SETFL 4
+#if defined(F_GETFL) && F_GETFL != ZVFS_F_GETFL
+#undef F_GETFL
+#endif
+#ifndef F_GETFL
+#define F_GETFL ZVFS_F_GETFL
+#endif
+
+#if defined(F_SETFL) && F_SETFL != ZVFS_F_SETFL
+#undef F_SETFL
+#endif
+#ifndef F_SETFL
+#define F_SETFL ZVFS_F_SETFL
+#endif

 /*
  * MWDT fcntl.h doesn't provide O_NONBLOCK, however it provides other file IO
  * definitions. Let's define O_NONBLOCK and check that it doesn't overlap with
  * any other file IO defines.
  */
+#if defined(O_NONBLOCK) && O_NONBLOCK != ZVFS_O_NONBLOCK
+#undef O_NONBLOCK
+#endif
 #ifndef O_NONBLOCK
-  #define O_NONBLOCK 0x4000
-  #if O_NONBLOCK & (O_RDONLY | O_WRONLY | O_RDWR | O_NDELAY | O_CREAT | O_APPEND | O_TRUNC | O_EXCL)
-    #error "O_NONBLOCK conflicts with other O_*** file IO defines!"
-  #endif
+  #define O_NONBLOCK ZVFS_O_NONBLOCK
+#endif
+#if O_NONBLOCK & (O_RDONLY | O_WRONLY | O_RDWR | O_NDELAY | O_CREAT | O_APPEND | O_TRUNC | O_EXCL)
+  #error "O_NONBLOCK conflicts with other O_*** file IO defines!"
 #endif

 #endif /* ZEPHYR_LIB_LIBC_ARCMWDT_INCLUDE_FCNTL_H_ */
diff --git a/lib/libc/armstdc/include/fcntl.h b/lib/libc/armstdc/include/fcntl.h
index 2341803e37a..c79fae36282 100644
--- a/lib/libc/armstdc/include/fcntl.h
+++ b/lib/libc/armstdc/include/fcntl.h
@@ -7,14 +7,16 @@
 #ifndef ZEPHYR_LIB_LIBC_ARMSTDC_INCLUDE_FCNTL_H_
 #define ZEPHYR_LIB_LIBC_ARMSTDC_INCLUDE_FCNTL_H_

+#include <zephyr/sys/fdtable.h>
+
 #define O_CREAT    0x0200
 #define O_APPEND   0x0400
 #define O_EXCL     0x0800
-#define O_NONBLOCK 0x4000
+#define O_NONBLOCK ZVFS_O_NONBLOCK

 #define F_DUPFD 0
-#define F_GETFL 3
-#define F_SETFL 4
+#define F_GETFL ZVFS_F_GETFL
+#define F_SETFL ZVFS_F_SETFL

 int open(const char *name, int flags, ...);

diff --git a/lib/libc/newlib/include/fcntl.h b/lib/libc/newlib/include/fcntl.h
new file mode 100644
index 00000000000..c817572f191
--- /dev/null
+++ b/lib/libc/newlib/include/fcntl.h
@@ -0,0 +1,35 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_FCNTL_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_FCNTL_H_
+
+#include_next <fcntl.h>
+
+#include <zephyr/sys/fdtable.h>
+
+#if defined(F_GETFL) && F_GETFL != ZVFS_F_GETFL
+#undef F_GETFL
+#endif
+#ifndef F_GETFL
+#define F_GETFL ZVFS_F_GETFL
+#endif
+
+#if defined(F_SETFL) && F_SETFL != ZVFS_F_SETFL
+#undef F_SETFL
+#endif
+#ifndef F_SETFL
+#define F_SETFL ZVFS_F_SETFL
+#endif
+
+#if defined(O_NONBLOCK) && O_NONBLOCK != ZVFS_O_NONBLOCK
+#undef O_NONBLOCK
+#endif
+#ifndef O_NONBLOCK
+#define O_NONBLOCK ZVFS_O_NONBLOCK
+#endif
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_FCNTL_H_ */
diff --git a/lib/libc/picolibc/include/fcntl.h b/lib/libc/picolibc/include/fcntl.h
new file mode 100644
index 00000000000..fe701829808
--- /dev/null
+++ b/lib/libc/picolibc/include/fcntl.h
@@ -0,0 +1,35 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_FCNTL_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_FCNTL_H_
+
+#include_next <fcntl.h>
+
+#include <zephyr/sys/fdtable.h>
+
+#if defined(F_GETFL) && F_GETFL != ZVFS_F_GETFL
+#undef F_GETFL
+#endif
+#ifndef F_GETFL
+#define F_GETFL ZVFS_F_GETFL
+#endif
+
+#if defined(F_SETFL) && F_SETFL != ZVFS_F_SETFL
+#undef F_SETFL
+#endif
+#ifndef F_SETFL
+#define F_SETFL ZVFS_F_SETFL
+#endif
+
+#if defined(O_NONBLOCK) && O_NONBLOCK != ZVFS_O_NONBLOCK
+#undef O_NONBLOCK
+#endif
+#ifndef O_NONBLOCK
+#define O_NONBLOCK ZVFS_O_NONBLOCK
+#endif
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_FCNTL_H_ */
diff --git a/samples/net/sockets/echo_service/src/main.c b/samples/net/sockets/echo_service/src/main.c
index 5f2b12c7298..85c64e52b64 100644
--- a/samples/net/sockets/echo_service/src/main.c
+++ b/samples/net/sockets/echo_service/src/main.c
@@ -4,6 +4,10 @@
  * SPDX-License-Identifier: Apache-2.0
  */

+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(net_echo_server_svc_sample, LOG_LEVEL_DBG);

diff --git a/subsys/net/lib/sockets/sockets.c b/subsys/net/lib/sockets/sockets.c
index 9c39eec31f9..17f38438cef 100644
--- a/subsys/net/lib/sockets/sockets.c
+++ b/subsys/net/lib/sockets/sockets.c
@@ -7,6 +7,21 @@
  */

 /* Zephyr headers */
+#include <fcntl.h>
+#include <zephyr/sys/fdtable.h>
+
+#ifndef F_GETFL
+#define F_GETFL ZVFS_F_GETFL
+#endif
+
+#ifndef F_SETFL
+#define F_SETFL ZVFS_F_SETFL
+#endif
+
+#ifndef O_NONBLOCK
+#define O_NONBLOCK ZVFS_O_NONBLOCK
+#endif
+
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(net_sock, CONFIG_NET_SOCKETS_LOG_LEVEL);

@@ -732,10 +747,30 @@ int z_impl_zsock_fcntl_impl(int sock, int cmd, int flags)
 	const struct socket_op_vtable *vtable;
 	struct k_mutex *lock;
 	void *obj;
+	int zvfs_cmd = cmd;
+	int zvfs_flags = flags;
 	int ret;

 	SYS_PORT_TRACING_OBJ_FUNC_ENTER(socket, fcntl, sock, cmd, flags);

+	if (cmd == F_GETFL) {
+#if F_GETFL != ZVFS_F_GETFL
+		LOG_WRN_ONCE("zsock_fcntl() called with F_GETFL; use ZVFS_F_GETFL instead");
+#endif
+		zvfs_cmd = ZVFS_F_GETFL;
+	} else if (cmd == F_SETFL) {
+#if F_SETFL != ZVFS_F_SETFL
+		LOG_WRN_ONCE("zsock_fcntl() called with F_SETFL; use ZVFS_F_SETFL instead");
+#endif
+		zvfs_cmd = ZVFS_F_SETFL;
+		if (flags & O_NONBLOCK) {
+#if O_NONBLOCK != ZVFS_O_NONBLOCK
+			LOG_WRN_ONCE("zsock_fcntl() called with O_NONBLOCK; use ZVFS_O_NONBLOCK instead");
+#endif
+			zvfs_flags |= ZVFS_O_NONBLOCK;
+		}
+	}
+
 	obj = get_sock_vtable(sock, &vtable, &lock);
 	if (obj == NULL) {
 		errno = EBADF;
@@ -746,7 +781,7 @@ int z_impl_zsock_fcntl_impl(int sock, int cmd, int flags)
 	(void)k_mutex_lock(lock, K_FOREVER);

 	ret = zvfs_fdtable_call_ioctl((const struct fd_op_vtable *)vtable,
-				   obj, cmd, flags);
+				   obj, zvfs_cmd, zvfs_flags);

 	k_mutex_unlock(lock);

diff --git a/tests/net/socket/tcp/src/main.c b/tests/net/socket/tcp/src/main.c
index 69ca1b7b6e3..6df83c12b49 100644
--- a/tests/net/socket/tcp/src/main.c
+++ b/tests/net/socket/tcp/src/main.c
@@ -4,6 +4,10 @@
  * SPDX-License-Identifier: Apache-2.0
  */

+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(net_test, CONFIG_NET_SOCKETS_LOG_LEVEL);

diff --git a/tests/net/socket/tls/src/main.c b/tests/net/socket/tls/src/main.c
index 5db92a18ff3..cb7ddb04041 100644
--- a/tests/net/socket/tls/src/main.c
+++ b/tests/net/socket/tls/src/main.c
@@ -4,6 +4,10 @@
  * SPDX-License-Identifier: Apache-2.0
  */

+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(net_test, CONFIG_NET_SOCKETS_LOG_LEVEL);

