diff --git a/lib/os/zvfs/zvfs_fdtable.c b/lib/os/zvfs/zvfs_fdtable.c
--- a/lib/os/zvfs/zvfs_fdtable.c
+++ b/lib/os/zvfs/zvfs_fdtable.c
@@ -881,6 +881,15 @@ static ssize_t stdinout_write_vmeth(void *obj, const void *buffer, size_t count)
 {
 #if defined(CONFIG_PICOLIBC) || defined(CONFIG_NEWLIB_LIBC) || defined(CONFIG_ARCMWDT_LIBC) || defined(CONFIG_MINIMAL_LIBC)
 	return z_impl_zephyr_write_stdout(buffer, count);
+#elif defined(CONFIG_PICOLIBC)
+	int z_impl_zephyr_fputc(int c, FILE *stream);
+	const char *cbuf = buffer;
+
+	for (size_t i = 0; i < count; i++) {
+		(void)z_impl_zephyr_fputc(cbuf[i], stdout);
+	}
+
+	return count;
 #else
 	return 0;
 #endif
