diff --git a/newlib/libc/tinystdio/CMakeLists.txt b/newlib/libc/tinystdio/CMakeLists.txt
index a992b35ef..5cd16b8e0 100644
--- a/newlib/libc/tinystdio/CMakeLists.txt
+++ b/newlib/libc/tinystdio/CMakeLists.txt
@@ -60,6 +60,7 @@ picolibc_sources(
   ferror.c
   fflush.c
   fgetc.c
+  fgetpos.c
   fgets.c
   fgetwc.c
   fgetws.c
@@ -69,6 +70,7 @@ picolibc_sources(
   filestrput.c
   filewstrget.c
   flockfile.c
+  flockfile_init.c
   fmemopen.c
   fopen.c
   fprintf.c
@@ -81,6 +83,7 @@ picolibc_sources(
   fscanf.c
   fseek.c
   fseeko.c
+  fsetpos.c
   ftell.c
   ftello.c
   ftox_engine.c
@@ -228,6 +231,8 @@ if(POSIX_CONSOLE)
     )
 endif()
 
+target_include_directories(c PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}")
+
 picolibc_headers(""
   stdio.h
   stdio-bufio.h
