diff --git a/src/utils/includes.h b/src/utils/includes.h
index 91784ecae..ae8b8d731 100644
--- a/src/utils/includes.h
+++ b/src/utils/includes.h
@@ -56,6 +56,13 @@
 #include <zephyr/net/socket.h>
 #include <zephyr/net/socket_select.h>
 #include <zephyr/net/net_ip.h>
+
+typedef zsock_fd_set fd_set;
+#define FD_ZERO ZSOCK_FD_ZERO
+#define FD_SET ZSOCK_FD_SET
+#define FD_CLR ZSOCK_FD_CLR
+#define FD_ISSET ZSOCK_FD_ISSET
+#define select zsock_select
 #endif /* defined(CONFIG_POSIX_API) */
 #include <zephyr/shell/shell.h>
 
